<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings SingleCellExperiment to the Tidyverse  • tidySingleCellExperiment</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings SingleCellExperiment to the Tidyverse ">
<meta property="og:description" content="tidySingleCellExperiment is an adapter that abstracts the SingleCellExperiment container 
    in the form of tibble and allows the data manipulation, plotting and nesting using tidyverse ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidySingleCellExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the tidySingleCellExperiment package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="tidysinglecellexperiment---part-of-tidytranscriptomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tidysinglecellexperiment---part-of-tidytranscriptomics" class="anchor"></a>tidySingleCellExperiment - part of tidytranscriptomics</h1></div>
<!-- badges: start -->

<p><strong>Brings SingleCellExperiment to the tidyverse!</strong></p>
<p>Website: <a href="https://stemangiola.github.io/tidySingleCellExperiment/articles/introduction.html">tidySingleCellExperiment</a></p>
<p>Please also have a look at</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidyseurat/">tidyseurat</a> for tidy manipulation of Seurat objects</li>
<li>
<a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> for tidy bulk RNA-seq data analysis</li>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>tidySingleCellExperiment provides a bridge between Bioconductor single-cell packages <br><span class="math display">@<em>a</em><em>m</em><em>e</em><em>z</em><em>q</em><em>u</em><em>i</em><em>t</em><em>a</em>2019<em>o</em><em>r</em><em>c</em><em>h</em><em>e</em><em>s</em><em>t</em><em>r</em><em>a</em><em>t</em><em>i</em><em>n</em><em>g</em></span><br> and the tidyverse <br><span class="math display">@<em>w</em><em>i</em><em>c</em><em>k</em><em>h</em><em>a</em><em>m</em>2019<em>w</em><em>e</em><em>l</em><em>c</em><em>o</em><em>m</em><em>e</em></span><br>. It creates an invisible layer that enables viewing the Bioconductor <em>SingleCellExperiment</em> object as a tidyverse tibble, and provides SingleCellExperiment-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions. This allows users to get the best of both Bioconductor and tidyverse worlds.</p>
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<colgroup>
<col width="39%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>SingleCellExperiment-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidySingleCellExperiment</code> is a SingleCellExperiment object, just better</td>
</tr></tbody>
</table>
<table class="table">
<colgroup>
<col width="24%">
<col width="75%">
</colgroup>
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> tibble functions (e.g. <code><a href="reference/arrange.html">tidySingleCellExperiment::select</a></code>)</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> tibble functions (e.g. <code><a href="reference/unnest.html">tidySingleCellExperiment::pivot_longer</a></code>)</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> (<code><a href="reference/ggplot.html">tidySingleCellExperiment::ggplot</a></code>)</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> (<code><a href="reference/plot_ly.html">tidySingleCellExperiment::plot_ly</a></code>)</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="23%">
<col width="76%">
</colgroup>
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidySingleCellExperiment</code> invisible layer over a SingleCellExperiment object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_transcripts</code></td>
<td>Add transcript-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From Bioconductor (under submission)</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly=</span><span class="ot">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2"></a>    <span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"tidySingleCellExperiment"</span>)</span></code></pre></div>
<p>From GitHub</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">"stemangiola/tidySingleCellExperiment"</span>)</span></code></pre></div>
<p>Load libraries used in this vignette.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># Bioconductor single-cell packages</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw">library</span>(scater)</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">library</span>(scran)</span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="kw">library</span>(SingleR)</span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="kw">library</span>(SingleCellSignalR)</span>
<span id="cb3-6"><a href="#cb3-6"></a></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co"># Tidyverse-compatible packages</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="kw">library</span>(purrr)</span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="kw">library</span>(tidyHeatmap)</span>
<span id="cb3-11"><a href="#cb3-11"></a></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co"># Both</span></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="kw">library</span>(tidySingleCellExperiment)</span></code></pre></div>
</div>
</div>
<div id="create-tidysinglecellexperiment-the-best-of-both-worlds" class="section level1">
<h1 class="hasAnchor">
<a href="#create-tidysinglecellexperiment-the-best-of-both-worlds" class="anchor"></a>Create <code>tidySingleCellExperiment</code>, the best of both worlds!</h1>
<p>This is a <em>SingleCellExperiment</em> object but it is evaluated as a tibble. So it is compatible both with SingleCellExperiment and tidyverse.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>pbmc_small_tidy &lt;-<span class="st"> </span>tidySingleCellExperiment<span class="op">::</span>pbmc_small <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">tidy</span>()</span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>pbmc_small_tidy</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">## # A tibble: 80 x 17</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb5-16"><a href="#cb5-16"></a><span class="co">## # … with 70 more rows, and 10 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="co">## #   ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;,</span></span>
<span id="cb5-18"><a href="#cb5-18"></a><span class="co">## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p><strong>But it is a SingleCellExperiment object after all</strong></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>pbmc_small_tidy<span class="op">@</span>assays</span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">## An object of class "SimpleAssays"</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">## Slot "data":</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">## List of length 2</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">## names(2): counts logcounts</span></span></code></pre></div>
</div>
<div id="annotation-polishing" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-polishing" class="anchor"></a>Annotation polishing</h1>
<p>We may have a column that contains the directory each run was taken from, such as the “file” column in <code>pbmc_small_tidy</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>pbmc_small_tidy<span class="op">$</span>file[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">## [1] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">## [2] "../data/sample1/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">## [3] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">## [4] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">## [5] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span></code></pre></div>
<p>We may want to extract the run/sample name out of it into a separate column. Tidyverse <code>extract</code> can be used to convert a character column into multiple columns using regular expression groups.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># Create sample column</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>pbmc_small_polished &lt;-</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="st">    </span>pbmc_small_tidy <span class="op">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="st">    </span><span class="kw">extract</span>(file, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, <span class="dt">remove=</span><span class="ot">FALSE</span>)</span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co"># Reorder to have sample column up front</span></span>
<span id="cb8-7"><a href="#cb8-7"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="st">    </span><span class="kw">select</span>(sample, <span class="kw">everything</span>())</span>
<span id="cb8-9"><a href="#cb8-9"></a></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="co">## # A tibble: 80 x 18</span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co">##    cell  sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;  &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        </span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="co">##  1 ATGC… sampl… SeuratPro…         70           47 0               A            </span></span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="co">##  2 CATG… sampl… SeuratPro…         85           52 0               A            </span></span>
<span id="cb8-15"><a href="#cb8-15"></a><span class="co">##  3 GAAC… sampl… SeuratPro…         87           50 1               B            </span></span>
<span id="cb8-16"><a href="#cb8-16"></a><span class="co">##  4 TGAC… sampl… SeuratPro…        127           56 0               A            </span></span>
<span id="cb8-17"><a href="#cb8-17"></a><span class="co">##  5 AGTC… sampl… SeuratPro…        173           53 0               A            </span></span>
<span id="cb8-18"><a href="#cb8-18"></a><span class="co">##  6 TCTG… sampl… SeuratPro…         70           48 0               A            </span></span>
<span id="cb8-19"><a href="#cb8-19"></a><span class="co">##  7 TGGT… sampl… SeuratPro…         64           36 0               A            </span></span>
<span id="cb8-20"><a href="#cb8-20"></a><span class="co">##  8 GCAG… sampl… SeuratPro…         72           45 0               A            </span></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="co">##  9 GATA… sampl… SeuratPro…         52           36 0               A            </span></span>
<span id="cb8-22"><a href="#cb8-22"></a><span class="co">## 10 AATG… sampl… SeuratPro…        100           41 0               A            </span></span>
<span id="cb8-23"><a href="#cb8-23"></a><span class="co">## # … with 70 more rows, and 11 more variables: groups &lt;chr&gt;,</span></span>
<span id="cb8-24"><a href="#cb8-24"></a><span class="co">## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,</span></span>
<span id="cb8-25"><a href="#cb8-25"></a><span class="co">## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div id="preliminary-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h1>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># Use colourblind-friendly colours</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>friendly_cols &lt;-<span class="st"> </span>dittoSeq<span class="op">::</span><span class="kw">dittoColors</span>()</span>
<span id="cb9-3"><a href="#cb9-3"></a></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co"># Set theme</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>my_theme &lt;-</span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="st">    </span><span class="kw">list</span>(</span>
<span id="cb9-7"><a href="#cb9-7"></a>        <span class="kw">scale_fill_manual</span>(<span class="dt">values=</span>friendly_cols),</span>
<span id="cb9-8"><a href="#cb9-8"></a>        <span class="kw">scale_color_manual</span>(<span class="dt">values=</span>friendly_cols),</span>
<span id="cb9-9"><a href="#cb9-9"></a>        <span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="st">            </span><span class="kw">theme</span>(</span>
<span id="cb9-11"><a href="#cb9-11"></a>                <span class="dt">panel.border=</span><span class="kw">element_blank</span>(),</span>
<span id="cb9-12"><a href="#cb9-12"></a>                <span class="dt">axis.line=</span><span class="kw">element_line</span>(),</span>
<span id="cb9-13"><a href="#cb9-13"></a>                <span class="dt">panel.grid.major=</span><span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.2</span>),</span>
<span id="cb9-14"><a href="#cb9-14"></a>                <span class="dt">panel.grid.minor=</span><span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.1</span>),</span>
<span id="cb9-15"><a href="#cb9-15"></a>                <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">12</span>),</span>
<span id="cb9-16"><a href="#cb9-16"></a>                <span class="dt">legend.position=</span><span class="st">"bottom"</span>,</span>
<span id="cb9-17"><a href="#cb9-17"></a>                <span class="dt">aspect.ratio=</span><span class="dv">1</span>,</span>
<span id="cb9-18"><a href="#cb9-18"></a>                <span class="dt">strip.background=</span><span class="kw">element_blank</span>(),</span>
<span id="cb9-19"><a href="#cb9-19"></a>                <span class="dt">axis.title.x=</span><span class="kw">element_text</span>(<span class="dt">margin=</span><span class="kw">margin</span>(<span class="dt">t=</span><span class="dv">10</span>, <span class="dt">r=</span><span class="dv">10</span>, <span class="dt">b=</span><span class="dv">10</span>, <span class="dt">l=</span><span class="dv">10</span>)),</span>
<span id="cb9-20"><a href="#cb9-20"></a>                <span class="dt">axis.title.y=</span><span class="kw">element_text</span>(<span class="dt">margin=</span><span class="kw">margin</span>(<span class="dt">t=</span><span class="dv">10</span>, <span class="dt">r=</span><span class="dv">10</span>, <span class="dt">b=</span><span class="dv">10</span>, <span class="dt">l=</span><span class="dv">10</span>))</span>
<span id="cb9-21"><a href="#cb9-21"></a>            )</span>
<span id="cb9-22"><a href="#cb9-22"></a>    )</span></code></pre></div>
<p>We can treat <code>pbmc_small_polished</code> as a tibble for plotting.</p>
<p>Here we plot number of transcripts per cell.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="st">    </span>tidySingleCellExperiment<span class="op">::</span><span class="kw">ggplot</span>(<span class="kw">aes</span>(nFeature_RNA, <span class="dt">fill=</span>groups)) <span class="op">+</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="st">    </span><span class="kw">geom_histogram</span>() <span class="op">+</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="st">    </span>my_theme</span>
<span id="cb10-5"><a href="#cb10-5"></a></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="reference/figures/plot1-1.png"><!-- --></p>
<p>Here we plot total transcripts per cell.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="st">    </span>tidySingleCellExperiment<span class="op">::</span><span class="kw">ggplot</span>(<span class="kw">aes</span>(groups, nCount_RNA, <span class="dt">fill=</span>groups)) <span class="op">+</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape=</span><span class="ot">NA</span>) <span class="op">+</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="st">    </span><span class="kw">geom_jitter</span>(<span class="dt">width=</span><span class="fl">0.1</span>) <span class="op">+</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="st">    </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/plot2-1.png"><!-- --></p>
<p>Here we plot abundance of two transcripts for each group.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="st">    </span><span class="kw">join_transcripts</span>(<span class="dt">transcripts=</span><span class="kw">c</span>(<span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(groups, abundance_counts <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">fill=</span>groups)) <span class="op">+</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape=</span><span class="ot">NA</span>) <span class="op">+</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="st">    </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">size=</span>nCount_RNA), <span class="dt">alpha=</span><span class="fl">0.5</span>, <span class="dt">width=</span><span class="fl">0.2</span>) <span class="op">+</span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="st">    </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="st">    </span>my_theme</span>
<span id="cb12-8"><a href="#cb12-8"></a></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-11-1.png"><!-- --></p>
</div>
<div id="preprocess-the-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h1>
<p>We can also treat <code>pbmc_small_polished</code> as a <em>SingleCellExperiment</em> object and proceed with data processing with Bioconductor packages, such as <em>scran</em> <br><span class="math display">@<em>l</em><em>u</em><em>n</em>2016<em>p</em><em>o</em><em>o</em><em>l</em><em>i</em><em>n</em><em>g</em></span><br> and <em>scater</em> <br><span class="math display">@<em>m</em><em>c</em><em>c</em><em>a</em><em>r</em><em>t</em><em>h</em><em>y</em>2017<em>s</em><em>c</em><em>a</em><em>t</em><em>e</em><em>r</em></span><br>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># Identify variable genes with scran</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>variable_genes &lt;-</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="st">    </span>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="st">    </span><span class="kw">modelGeneVar</span>() <span class="op">%&gt;%</span></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="st">    </span><span class="kw">getTopHVGs</span>(<span class="dt">prop=</span><span class="fl">0.1</span>)</span>
<span id="cb13-6"><a href="#cb13-6"></a></span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="co"># Perform PCA with scater</span></span>
<span id="cb13-8"><a href="#cb13-8"></a>pbmc_small_pca &lt;-</span>
<span id="cb13-9"><a href="#cb13-9"></a><span class="st">    </span>pbmc_small_polished <span class="op">%&gt;%</span></span>
<span id="cb13-10"><a href="#cb13-10"></a><span class="st">    </span><span class="kw">runPCA</span>(<span class="dt">subset_row=</span>variable_genes)</span>
<span id="cb13-11"><a href="#cb13-11"></a></span>
<span id="cb13-12"><a href="#cb13-12"></a><span class="co">## Warning in check_numbers(k = k, nu = nu, nv = nv, limit = min(dim(x)) - : more</span></span>
<span id="cb13-13"><a href="#cb13-13"></a><span class="co">## singular values/vectors requested than available</span></span>
<span id="cb13-14"><a href="#cb13-14"></a></span>
<span id="cb13-15"><a href="#cb13-15"></a><span class="co">## Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =</span></span>
<span id="cb13-16"><a href="#cb13-16"></a><span class="co">## TRUE, : You're computing too large a percentage of total singular values, use a</span></span>
<span id="cb13-17"><a href="#cb13-17"></a><span class="co">## standard svd instead.</span></span>
<span id="cb13-18"><a href="#cb13-18"></a></span>
<span id="cb13-19"><a href="#cb13-19"></a>pbmc_small_pca</span>
<span id="cb13-20"><a href="#cb13-20"></a></span>
<span id="cb13-21"><a href="#cb13-21"></a><span class="co">## # A tibble: 80 x 18</span></span>
<span id="cb13-22"><a href="#cb13-22"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb13-23"><a href="#cb13-23"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb13-24"><a href="#cb13-24"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb13-25"><a href="#cb13-25"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb13-26"><a href="#cb13-26"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb13-27"><a href="#cb13-27"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb13-28"><a href="#cb13-28"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb13-29"><a href="#cb13-29"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb13-30"><a href="#cb13-30"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb13-31"><a href="#cb13-31"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb13-32"><a href="#cb13-32"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb13-33"><a href="#cb13-33"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb13-34"><a href="#cb13-34"></a><span class="co">## # … with 70 more rows, and 11 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb13-35"><a href="#cb13-35"></a><span class="co">## #   sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;,</span></span>
<span id="cb13-36"><a href="#cb13-36"></a><span class="co">## #   PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>If a tidyverse-compatible package is not included in the tidySingleCellExperiment collection, we can use <code>as_tibble</code> to permanently convert <code>tidySingleCellExperiment</code> into a tibble.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># Create pairs plot with GGally</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>pbmc_small_pca <span class="op">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="st">    </span><span class="kw">select</span>(<span class="kw">contains</span>(<span class="st">"PC"</span>), <span class="kw">everything</span>()) <span class="op">%&gt;%</span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="st">    </span>GGally<span class="op">::</span><span class="kw">ggpairs</span>(<span class="dt">columns=</span><span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ggplot2<span class="op">::</span><span class="kw">aes</span>(<span class="dt">colour=</span>groups)) <span class="op">+</span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="st">    </span>my_theme</span>
<span id="cb14-7"><a href="#cb14-7"></a></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">##   method from   </span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">##   +.gg   ggplot2</span></span></code></pre></div>
<p><img src="reference/figures/pc_plot-1.png"><!-- --></p>
</div>
<div id="identify-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h1>
<p>We can proceed with cluster identification with <em>scran</em>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>pbmc_small_cluster &lt;-<span class="st"> </span>pbmc_small_pca</span>
<span id="cb15-2"><a href="#cb15-2"></a></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="co"># Assign clusters to the 'colLabels' of the SummarizedExperiment object</span></span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="kw">colLabels</span>(pbmc_small_cluster) &lt;-</span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="st">    </span>pbmc_small_pca <span class="op">%&gt;%</span></span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="st">    </span><span class="kw">buildSNNGraph</span>(<span class="dt">use.dimred=</span><span class="st">"PCA"</span>) <span class="op">%&gt;%</span></span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="st">    </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>() <span class="op">%$%</span></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="st">    </span>membership <span class="op">%&gt;%</span></span>
<span id="cb15-9"><a href="#cb15-9"></a><span class="st">    </span><span class="kw">as.factor</span>()</span>
<span id="cb15-10"><a href="#cb15-10"></a></span>
<span id="cb15-11"><a href="#cb15-11"></a><span class="co">## Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span id="cb15-12"><a href="#cb15-12"></a><span class="co">## detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span>
<span id="cb15-13"><a href="#cb15-13"></a></span>
<span id="cb15-14"><a href="#cb15-14"></a><span class="co"># Reorder columns</span></span>
<span id="cb15-15"><a href="#cb15-15"></a>pbmc_small_cluster <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(label, <span class="kw">everything</span>())</span>
<span id="cb15-16"><a href="#cb15-16"></a></span>
<span id="cb15-17"><a href="#cb15-17"></a><span class="co">## # A tibble: 80 x 19</span></span>
<span id="cb15-18"><a href="#cb15-18"></a><span class="co">##    cell  label orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb15-19"><a href="#cb15-19"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        </span></span>
<span id="cb15-20"><a href="#cb15-20"></a><span class="co">##  1 ATGC… 2     SeuratPro…         70           47 0               A            </span></span>
<span id="cb15-21"><a href="#cb15-21"></a><span class="co">##  2 CATG… 2     SeuratPro…         85           52 0               A            </span></span>
<span id="cb15-22"><a href="#cb15-22"></a><span class="co">##  3 GAAC… 2     SeuratPro…         87           50 1               B            </span></span>
<span id="cb15-23"><a href="#cb15-23"></a><span class="co">##  4 TGAC… 1     SeuratPro…        127           56 0               A            </span></span>
<span id="cb15-24"><a href="#cb15-24"></a><span class="co">##  5 AGTC… 2     SeuratPro…        173           53 0               A            </span></span>
<span id="cb15-25"><a href="#cb15-25"></a><span class="co">##  6 TCTG… 2     SeuratPro…         70           48 0               A            </span></span>
<span id="cb15-26"><a href="#cb15-26"></a><span class="co">##  7 TGGT… 1     SeuratPro…         64           36 0               A            </span></span>
<span id="cb15-27"><a href="#cb15-27"></a><span class="co">##  8 GCAG… 2     SeuratPro…         72           45 0               A            </span></span>
<span id="cb15-28"><a href="#cb15-28"></a><span class="co">##  9 GATA… 2     SeuratPro…         52           36 0               A            </span></span>
<span id="cb15-29"><a href="#cb15-29"></a><span class="co">## 10 AATG… 2     SeuratPro…        100           41 0               A            </span></span>
<span id="cb15-30"><a href="#cb15-30"></a><span class="co">## # … with 70 more rows, and 12 more variables: groups &lt;chr&gt;,</span></span>
<span id="cb15-31"><a href="#cb15-31"></a><span class="co">## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;,</span></span>
<span id="cb15-32"><a href="#cb15-32"></a><span class="co">## #   PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>And interrogate the output as if it was a regular tibble.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># Count number of cells for each cluster per group</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="st">    </span>tidySingleCellExperiment<span class="op">::</span><span class="kw">count</span>(groups, label)</span>
<span id="cb16-4"><a href="#cb16-4"></a></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="co">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span>
<span id="cb16-6"><a href="#cb16-6"></a></span>
<span id="cb16-7"><a href="#cb16-7"></a><span class="co">## # A tibble: 8 x 3</span></span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="co">##   groups label     n</span></span>
<span id="cb16-9"><a href="#cb16-9"></a><span class="co">##   &lt;chr&gt;  &lt;fct&gt; &lt;int&gt;</span></span>
<span id="cb16-10"><a href="#cb16-10"></a><span class="co">## 1 g1     1        12</span></span>
<span id="cb16-11"><a href="#cb16-11"></a><span class="co">## 2 g1     2        14</span></span>
<span id="cb16-12"><a href="#cb16-12"></a><span class="co">## 3 g1     3        14</span></span>
<span id="cb16-13"><a href="#cb16-13"></a><span class="co">## 4 g1     4         4</span></span>
<span id="cb16-14"><a href="#cb16-14"></a><span class="co">## 5 g2     1        10</span></span>
<span id="cb16-15"><a href="#cb16-15"></a><span class="co">## 6 g2     2        11</span></span>
<span id="cb16-16"><a href="#cb16-16"></a><span class="co">## 7 g2     3        10</span></span>
<span id="cb16-17"><a href="#cb16-17"></a><span class="co">## 8 g2     4         5</span></span></code></pre></div>
<p>We can identify and visualise cluster markers combining SingleCellExperiment, tidyverse functions and tidyHeatmap <br><span class="math display">@<em>m</em><em>a</em><em>n</em><em>g</em><em>i</em><em>o</em><em>l</em><em>a</em>2020<em>t</em><em>i</em><em>d</em><em>y</em><em>h</em><em>e</em><em>a</em><em>t</em><em>m</em><em>a</em><em>p</em></span><br></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># Identify top 10 markers per cluster</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>marker_genes &lt;-</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="st">    </span>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="st">    </span><span class="kw">findMarkers</span>(<span class="dt">groups=</span>pbmc_small_cluster<span class="op">$</span>label) <span class="op">%&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="st">    </span><span class="kw">as.list</span>() <span class="op">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="st">    </span><span class="kw">map</span>(<span class="op">~</span><span class="st"> </span>.x <span class="op">%&gt;%</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="st">        </span><span class="kw">head</span>(<span class="dv">10</span>) <span class="op">%&gt;%</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="st">        </span><span class="kw">rownames</span>()) <span class="op">%&gt;%</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="st">    </span><span class="kw">unlist</span>()</span>
<span id="cb17-10"><a href="#cb17-10"></a></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co"># Plot heatmap</span></span>
<span id="cb17-12"><a href="#cb17-12"></a>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="st">    </span><span class="kw">join_transcripts</span>(<span class="dt">transcripts=</span>marker_genes) <span class="op">%&gt;%</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="st">    </span><span class="kw">group_by</span>(label) <span class="op">%&gt;%</span></span>
<span id="cb17-15"><a href="#cb17-15"></a><span class="st">    </span><span class="kw">heatmap</span>(transcript, cell, abundance_counts, <span class="dt">.scale=</span><span class="st">"column"</span>)</span>
<span id="cb17-16"><a href="#cb17-16"></a></span>
<span id="cb17-17"><a href="#cb17-17"></a><span class="co">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-12-1.png"><!-- --></p>
</div>
<div id="reduce-dimensions" class="section level1">
<h1 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h1>
<p>We can calculate the first 3 UMAP dimensions using the SingleCellExperiment framework and <em>scater</em>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>pbmc_small_UMAP &lt;-</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">    </span>pbmc_small_cluster <span class="op">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="st">    </span><span class="kw">runUMAP</span>(<span class="dt">ncomponents=</span><span class="dv">3</span>)</span></code></pre></div>
<p>And we can plot the result in 3D using plotly.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>pbmc_small_UMAP <span class="op">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="st">    </span><span class="kw">plot_ly</span>(</span>
<span id="cb19-3"><a href="#cb19-3"></a>        <span class="dt">x=</span><span class="op">~</span><span class="st">`</span><span class="dt">UMAP1</span><span class="st">`</span>,</span>
<span id="cb19-4"><a href="#cb19-4"></a>        <span class="dt">y=</span><span class="op">~</span><span class="st">`</span><span class="dt">UMAP2</span><span class="st">`</span>,</span>
<span id="cb19-5"><a href="#cb19-5"></a>        <span class="dt">z=</span><span class="op">~</span><span class="st">`</span><span class="dt">UMAP3</span><span class="st">`</span>,</span>
<span id="cb19-6"><a href="#cb19-6"></a>        <span class="dt">color=</span><span class="op">~</span>label,</span>
<span id="cb19-7"><a href="#cb19-7"></a>        <span class="dt">colors=</span>friendly_cols[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</span>
<span id="cb19-8"><a href="#cb19-8"></a>    )</span></code></pre></div>
<p><img src="reference/figures/plotly.png" alt="plotly screenshot"></p>
</div>
<div id="cell-type-prediction" class="section level1">
<h1 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h1>
<p>We can infer cell type identities using <em>SingleR</em> <br><span class="math display">@<em>a</em><em>r</em><em>a</em><em>n</em>2019<em>r</em><em>e</em><em>f</em><em>e</em><em>r</em><em>e</em><em>n</em><em>c</em><em>e</em></span><br> and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="co"># Get cell type reference data</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>blueprint &lt;-<span class="st"> </span>celldex<span class="op">::</span><span class="kw">BlueprintEncodeData</span>()</span>
<span id="cb20-3"><a href="#cb20-3"></a></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="co">## snapshotDate(): 2020-09-04</span></span>
<span id="cb20-5"><a href="#cb20-5"></a></span>
<span id="cb20-6"><a href="#cb20-6"></a><span class="co">## see ?celldex and browseVignettes('celldex') for documentation</span></span>
<span id="cb20-7"><a href="#cb20-7"></a></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="co">## loading from cache</span></span>
<span id="cb20-9"><a href="#cb20-9"></a></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="co">## see ?celldex and browseVignettes('celldex') for documentation</span></span>
<span id="cb20-11"><a href="#cb20-11"></a></span>
<span id="cb20-12"><a href="#cb20-12"></a><span class="co">## loading from cache</span></span>
<span id="cb20-13"><a href="#cb20-13"></a></span>
<span id="cb20-14"><a href="#cb20-14"></a><span class="co"># Infer cell identities</span></span>
<span id="cb20-15"><a href="#cb20-15"></a>cell_type_df &lt;-</span>
<span id="cb20-16"><a href="#cb20-16"></a><span class="st">    </span>pbmc_small_UMAP<span class="op">@</span>assays<span class="op">@</span>data<span class="op">$</span>logcounts <span class="op">%&gt;%</span></span>
<span id="cb20-17"><a href="#cb20-17"></a><span class="st">    </span>Matrix<span class="op">::</span><span class="kw">Matrix</span>(<span class="dt">sparse=</span><span class="ot">TRUE</span>) <span class="op">%&gt;%</span></span>
<span id="cb20-18"><a href="#cb20-18"></a><span class="st">    </span><span class="kw">SingleR</span>(</span>
<span id="cb20-19"><a href="#cb20-19"></a>        <span class="dt">ref=</span>blueprint,</span>
<span id="cb20-20"><a href="#cb20-20"></a>        <span class="dt">labels=</span>blueprint<span class="op">$</span>label.main</span>
<span id="cb20-21"><a href="#cb20-21"></a>    ) <span class="op">%&gt;%</span></span>
<span id="cb20-22"><a href="#cb20-22"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span></span>
<span id="cb20-23"><a href="#cb20-23"></a><span class="st">    </span><span class="kw">as_tibble</span>(<span class="dt">rownames=</span><span class="st">"cell"</span>) <span class="op">%&gt;%</span></span>
<span id="cb20-24"><a href="#cb20-24"></a><span class="st">    </span><span class="kw">select</span>(cell, first.labels)</span>
<span id="cb20-25"><a href="#cb20-25"></a></span>
<span id="cb20-26"><a href="#cb20-26"></a><span class="co"># Join UMAP and cell type info</span></span>
<span id="cb20-27"><a href="#cb20-27"></a>pbmc_small_cell_type &lt;-</span>
<span id="cb20-28"><a href="#cb20-28"></a><span class="st">    </span>pbmc_small_UMAP <span class="op">%&gt;%</span></span>
<span id="cb20-29"><a href="#cb20-29"></a><span class="st">    </span><span class="kw">left_join</span>(cell_type_df, <span class="dt">by=</span><span class="st">"cell"</span>)</span>
<span id="cb20-30"><a href="#cb20-30"></a></span>
<span id="cb20-31"><a href="#cb20-31"></a><span class="co"># Reorder columns</span></span>
<span id="cb20-32"><a href="#cb20-32"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb20-33"><a href="#cb20-33"></a><span class="st">    </span>tidySingleCellExperiment<span class="op">::</span><span class="kw">select</span>(cell, first.labels, <span class="kw">everything</span>())</span>
<span id="cb20-34"><a href="#cb20-34"></a></span>
<span id="cb20-35"><a href="#cb20-35"></a><span class="co">## # A tibble: 80 x 23</span></span>
<span id="cb20-36"><a href="#cb20-36"></a><span class="co">##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span id="cb20-37"><a href="#cb20-37"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          </span></span>
<span id="cb20-38"><a href="#cb20-38"></a><span class="co">##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              </span></span>
<span id="cb20-39"><a href="#cb20-39"></a><span class="co">##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              </span></span>
<span id="cb20-40"><a href="#cb20-40"></a><span class="co">##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              </span></span>
<span id="cb20-41"><a href="#cb20-41"></a><span class="co">##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              </span></span>
<span id="cb20-42"><a href="#cb20-42"></a><span class="co">##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              </span></span>
<span id="cb20-43"><a href="#cb20-43"></a><span class="co">##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              </span></span>
<span id="cb20-44"><a href="#cb20-44"></a><span class="co">##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              </span></span>
<span id="cb20-45"><a href="#cb20-45"></a><span class="co">##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              </span></span>
<span id="cb20-46"><a href="#cb20-46"></a><span class="co">##  9 GATA… CD8+ T-cells SeuratPro…         52           36 0              </span></span>
<span id="cb20-47"><a href="#cb20-47"></a><span class="co">## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              </span></span>
<span id="cb20-48"><a href="#cb20-48"></a><span class="co">## # … with 70 more rows, and 17 more variables: letter.idents &lt;fct&gt;,</span></span>
<span id="cb20-49"><a href="#cb20-49"></a><span class="co">## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;,</span></span>
<span id="cb20-50"><a href="#cb20-50"></a><span class="co">## #   label &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;,</span></span>
<span id="cb20-51"><a href="#cb20-51"></a><span class="co">## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;</span></span></code></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co"># Count number of cells for each cell type per cluster</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="st">    </span><span class="kw">count</span>(label, first.labels)</span>
<span id="cb21-4"><a href="#cb21-4"></a></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="co">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span>
<span id="cb21-6"><a href="#cb21-6"></a></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="co">## # A tibble: 9 x 3</span></span>
<span id="cb21-8"><a href="#cb21-8"></a><span class="co">##   label first.labels     n</span></span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="co">##   &lt;fct&gt; &lt;chr&gt;        &lt;int&gt;</span></span>
<span id="cb21-10"><a href="#cb21-10"></a><span class="co">## 1 1     CD4+ T-cells     2</span></span>
<span id="cb21-11"><a href="#cb21-11"></a><span class="co">## 2 1     CD8+ T-cells     8</span></span>
<span id="cb21-12"><a href="#cb21-12"></a><span class="co">## 3 1     NK cells        12</span></span>
<span id="cb21-13"><a href="#cb21-13"></a><span class="co">## 4 2     B-cells         10</span></span>
<span id="cb21-14"><a href="#cb21-14"></a><span class="co">## 5 2     CD4+ T-cells     5</span></span>
<span id="cb21-15"><a href="#cb21-15"></a><span class="co">## 6 2     CD8+ T-cells     3</span></span>
<span id="cb21-16"><a href="#cb21-16"></a><span class="co">## 7 2     Monocytes        7</span></span>
<span id="cb21-17"><a href="#cb21-17"></a><span class="co">## 8 3     Monocytes       24</span></span>
<span id="cb21-18"><a href="#cb21-18"></a><span class="co">## 9 4     Erythrocytes     9</span></span></code></pre></div>
<p>We can easily reshape the data for building information-rich faceted plots.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="st">    </span><span class="co"># Reshape and add classifier column</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="st">    </span><span class="kw">pivot_longer</span>(</span>
<span id="cb22-5"><a href="#cb22-5"></a>        <span class="dt">cols=</span><span class="kw">c</span>(label, first.labels),</span>
<span id="cb22-6"><a href="#cb22-6"></a>        <span class="dt">names_to=</span><span class="st">"classifier"</span>, <span class="dt">values_to=</span><span class="st">"label"</span></span>
<span id="cb22-7"><a href="#cb22-7"></a>    ) <span class="op">%&gt;%</span></span>
<span id="cb22-8"><a href="#cb22-8"></a></span>
<span id="cb22-9"><a href="#cb22-9"></a><span class="st">    </span><span class="co"># UMAP plots for cell type and cluster</span></span>
<span id="cb22-10"><a href="#cb22-10"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(UMAP1, UMAP2, <span class="dt">color=</span>label)) <span class="op">+</span></span>
<span id="cb22-11"><a href="#cb22-11"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb22-12"><a href="#cb22-12"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>classifier) <span class="op">+</span></span>
<span id="cb22-13"><a href="#cb22-13"></a><span class="st">    </span>my_theme</span>
<span id="cb22-14"><a href="#cb22-14"></a></span>
<span id="cb22-15"><a href="#cb22-15"></a><span class="co">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-15-1.png"><!-- --></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb23-2"><a href="#cb23-2"></a></span>
<span id="cb23-3"><a href="#cb23-3"></a><span class="st">    </span><span class="co"># Add some mitochondrial abundance values</span></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">mitochondrial=</span><span class="kw">rnorm</span>(dplyr<span class="op">::</span><span class="kw">n</span>())) <span class="op">%&gt;%</span></span>
<span id="cb23-5"><a href="#cb23-5"></a></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="st">    </span><span class="co"># Plot correlation</span></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="st">    </span><span class="kw">join_transcripts</span>(<span class="dt">transcripts=</span><span class="kw">c</span>(<span class="st">"CST3"</span>, <span class="st">"LYZ"</span>), <span class="dt">shape=</span><span class="st">"wide"</span>) <span class="op">%&gt;%</span></span>
<span id="cb23-8"><a href="#cb23-8"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(CST3 <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, LYZ <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">color=</span>groups, <span class="dt">size=</span>mitochondrial)) <span class="op">+</span></span>
<span id="cb23-9"><a href="#cb23-9"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb23-10"><a href="#cb23-10"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>first.labels, <span class="dt">scales=</span><span class="st">"free"</span>) <span class="op">+</span></span>
<span id="cb23-11"><a href="#cb23-11"></a><span class="st">    </span><span class="kw">scale_x_log10</span>() <span class="op">+</span></span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="st">    </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb23-13"><a href="#cb23-13"></a><span class="st">    </span>my_theme</span>
<span id="cb23-14"><a href="#cb23-14"></a></span>
<span id="cb23-15"><a href="#cb23-15"></a><span class="co">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-16-1.png"><!-- --></p>
</div>
<div id="nested-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h1>
<p>A powerful tool we can use with tidySingleCellExperiment is tidyverse <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types into lymphoid and myeloid, and then nest based on the new classification.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>pbmc_small_nested &lt;-</span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="st">    </span>pbmc_small_cell_type <span class="op">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="st">    </span><span class="kw">filter</span>(first.labels <span class="op">!=</span><span class="st"> "Erythrocytes"</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">cell_class=</span>dplyr<span class="op">::</span><span class="kw">if_else</span>(<span class="st">`</span><span class="dt">first.labels</span><span class="st">`</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span>), <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="st">    </span><span class="kw">nest</span>(<span class="dt">data=</span><span class="op">-</span>cell_class)</span>
<span id="cb24-6"><a href="#cb24-6"></a></span>
<span id="cb24-7"><a href="#cb24-7"></a>pbmc_small_nested</span>
<span id="cb24-8"><a href="#cb24-8"></a></span>
<span id="cb24-9"><a href="#cb24-9"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb24-10"><a href="#cb24-10"></a><span class="co">##   cell_class data     </span></span>
<span id="cb24-11"><a href="#cb24-11"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;   </span></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="co">## 1 lymphoid   &lt;tidySingleCellExperiment&gt;</span></span>
<span id="cb24-13"><a href="#cb24-13"></a><span class="co">## 2 myeloid    &lt;tidySingleCellExperiment&gt;</span></span></code></pre></div>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>pbmc_small_nested_reanalysed &lt;-</span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="st">    </span>pbmc_small_nested <span class="op">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">data=</span><span class="kw">map</span>(</span>
<span id="cb25-4"><a href="#cb25-4"></a>        data, <span class="op">~</span><span class="st"> </span>{</span>
<span id="cb25-5"><a href="#cb25-5"></a>            .x &lt;-<span class="st"> </span><span class="kw">runPCA</span>(.x, <span class="dt">subset_row=</span>variable_genes)</span>
<span id="cb25-6"><a href="#cb25-6"></a></span>
<span id="cb25-7"><a href="#cb25-7"></a>            variable_genes &lt;-</span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="st">                </span>.x <span class="op">%&gt;%</span></span>
<span id="cb25-9"><a href="#cb25-9"></a><span class="st">                </span><span class="kw">modelGeneVar</span>() <span class="op">%&gt;%</span></span>
<span id="cb25-10"><a href="#cb25-10"></a><span class="st">                </span><span class="kw">getTopHVGs</span>(<span class="dt">prop=</span><span class="fl">0.3</span>)</span>
<span id="cb25-11"><a href="#cb25-11"></a></span>
<span id="cb25-12"><a href="#cb25-12"></a>            <span class="kw">colLabels</span>(.x) &lt;-</span>
<span id="cb25-13"><a href="#cb25-13"></a><span class="st">                </span>.x <span class="op">%&gt;%</span></span>
<span id="cb25-14"><a href="#cb25-14"></a><span class="st">                </span><span class="kw">buildSNNGraph</span>(<span class="dt">use.dimred=</span><span class="st">"PCA"</span>) <span class="op">%&gt;%</span></span>
<span id="cb25-15"><a href="#cb25-15"></a><span class="st">                </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>() <span class="op">%$%</span></span>
<span id="cb25-16"><a href="#cb25-16"></a><span class="st">                </span>membership <span class="op">%&gt;%</span></span>
<span id="cb25-17"><a href="#cb25-17"></a><span class="st">                </span><span class="kw">as.factor</span>()</span>
<span id="cb25-18"><a href="#cb25-18"></a></span>
<span id="cb25-19"><a href="#cb25-19"></a>            .x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">runUMAP</span>(<span class="dt">ncomponents=</span><span class="dv">3</span>)</span>
<span id="cb25-20"><a href="#cb25-20"></a>        }</span>
<span id="cb25-21"><a href="#cb25-21"></a>    ))</span>
<span id="cb25-22"><a href="#cb25-22"></a></span>
<span id="cb25-23"><a href="#cb25-23"></a>pbmc_small_nested_reanalysed</span>
<span id="cb25-24"><a href="#cb25-24"></a></span>
<span id="cb25-25"><a href="#cb25-25"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb25-26"><a href="#cb25-26"></a><span class="co">##   cell_class data     </span></span>
<span id="cb25-27"><a href="#cb25-27"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;   </span></span>
<span id="cb25-28"><a href="#cb25-28"></a><span class="co">## 1 lymphoid   &lt;tidySingleCellExperiment&gt;</span></span>
<span id="cb25-29"><a href="#cb25-29"></a><span class="co">## 2 myeloid    &lt;tidySingleCellExperiment&gt;</span></span></code></pre></div>
<p>We can then unnest and plot the new classification.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>pbmc_small_nested_reanalysed <span class="op">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="st">    </span><span class="co"># Convert to tibble otherwise SingleCellExperiment drops reduced dimensions when unifying data sets.</span></span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">data=</span><span class="kw">map</span>(data, <span class="op">~</span><span class="st"> </span>.x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>())) <span class="op">%&gt;%</span></span>
<span id="cb26-5"><a href="#cb26-5"></a><span class="st">    </span><span class="kw">unnest</span>(data) <span class="op">%&gt;%</span></span>
<span id="cb26-6"><a href="#cb26-6"></a></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="st">    </span><span class="co"># Define unique clusters</span></span>
<span id="cb26-8"><a href="#cb26-8"></a><span class="st">    </span><span class="kw">unite</span>(<span class="st">"cluster"</span>, <span class="kw">c</span>(cell_class, label), <span class="dt">remove=</span><span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb26-9"><a href="#cb26-9"></a></span>
<span id="cb26-10"><a href="#cb26-10"></a><span class="st">    </span><span class="co"># Plotting</span></span>
<span id="cb26-11"><a href="#cb26-11"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(UMAP1, UMAP2, <span class="dt">color=</span>cluster)) <span class="op">+</span></span>
<span id="cb26-12"><a href="#cb26-12"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb26-13"><a href="#cb26-13"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>cell_class) <span class="op">+</span></span>
<span id="cb26-14"><a href="#cb26-14"></a><span class="st">    </span>my_theme</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-19-1.png"><!-- --></p>
<p>We can perform a large number of functional analyses on data subsets. For example, we can identify intra-sample cell-cell interactions using <em>SingleCellSignalR</em> <br><span class="math display">@<em>c</em><em>a</em><em>b</em><em>e</em><em>l</em><em>l</em><em>o</em>2020<em>s</em><em>i</em><em>n</em><em>g</em><em>l</em><em>e</em><em>c</em><em>e</em><em>l</em><em>l</em><em>s</em><em>i</em><em>g</em><em>n</em><em>a</em><em>l</em><em>r</em></span><br>, and then compare whether interactions are stronger or weaker across conditions. The code below demonstrates how this analysis could be performed. It won’t work with this small example dataset as we have just two samples (one for each condition). But some example output is shown below and you can imagine how you can use tidyverse on the output to perform t-tests and visualisation.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>pbmc_small_nested_interactions &lt;-</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="st">    </span>pbmc_small_nested_reanalysed <span class="op">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3"></a></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="st">    </span><span class="co"># Unnest based on cell category</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="st">    </span><span class="kw">unnest</span>(data) <span class="op">%&gt;%</span></span>
<span id="cb27-6"><a href="#cb27-6"></a></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="st">    </span><span class="co"># Create unambiguous clusters</span></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">integrated_clusters=</span>first.labels <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.factor</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.integer</span>()) <span class="op">%&gt;%</span></span>
<span id="cb27-9"><a href="#cb27-9"></a></span>
<span id="cb27-10"><a href="#cb27-10"></a><span class="st">    </span><span class="co"># Nest based on sample</span></span>
<span id="cb27-11"><a href="#cb27-11"></a><span class="st">    </span>tidySingleCellExperiment<span class="op">::</span><span class="kw">nest</span>(<span class="dt">data=</span><span class="op">-</span>sample) <span class="op">%&gt;%</span></span>
<span id="cb27-12"><a href="#cb27-12"></a><span class="st">    </span>tidySingleCellExperiment<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">interactions=</span><span class="kw">map</span>(data, <span class="op">~</span><span class="st"> </span>{</span>
<span id="cb27-13"><a href="#cb27-13"></a></span>
<span id="cb27-14"><a href="#cb27-14"></a>        <span class="co"># Produce variables. Yuck!</span></span>
<span id="cb27-15"><a href="#cb27-15"></a>        cluster &lt;-<span class="st"> </span>.x<span class="op">@</span>colData<span class="op">$</span>integrated_clusters</span>
<span id="cb27-16"><a href="#cb27-16"></a>        data &lt;-<span class="st"> </span><span class="kw">data.frame</span>(.x<span class="op">@</span>assays<span class="op">@</span>data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.list</span>() <span class="op">%&gt;%</span><span class="st"> </span>.[[<span class="dv">1</span>]] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.matrix</span>())</span>
<span id="cb27-17"><a href="#cb27-17"></a></span>
<span id="cb27-18"><a href="#cb27-18"></a>        <span class="co"># Ligand/Receptor analysis using SingleCellSignalR</span></span>
<span id="cb27-19"><a href="#cb27-19"></a>        data <span class="op">%&gt;%</span></span>
<span id="cb27-20"><a href="#cb27-20"></a><span class="st">            </span><span class="kw">cell_signaling</span>(<span class="dt">genes=</span><span class="kw">rownames</span>(data), <span class="dt">cluster=</span>cluster) <span class="op">%&gt;%</span></span>
<span id="cb27-21"><a href="#cb27-21"></a><span class="st">            </span><span class="kw">inter_network</span>(<span class="dt">data=</span>data, <span class="dt">signal=</span>., <span class="dt">genes=</span><span class="kw">rownames</span>(data), <span class="dt">cluster=</span>cluster) <span class="op">%$%</span></span>
<span id="cb27-22"><a href="#cb27-22"></a><span class="st">            `</span><span class="dt">individual-networks</span><span class="st">`</span> <span class="op">%&gt;%</span></span>
<span id="cb27-23"><a href="#cb27-23"></a><span class="st">            </span><span class="kw">map_dfr</span>(<span class="op">~</span><span class="st"> </span><span class="kw">bind_rows</span>(<span class="kw">as_tibble</span>(.x)))</span>
<span id="cb27-24"><a href="#cb27-24"></a>    }))</span>
<span id="cb27-25"><a href="#cb27-25"></a></span>
<span id="cb27-26"><a href="#cb27-26"></a>pbmc_small_nested_interactions <span class="op">%&gt;%</span></span>
<span id="cb27-27"><a href="#cb27-27"></a><span class="st">    </span><span class="kw">select</span>(<span class="op">-</span>data) <span class="op">%&gt;%</span></span>
<span id="cb27-28"><a href="#cb27-28"></a><span class="st">    </span><span class="kw">unnest</span>(interactions)</span></code></pre></div>
<p>If the dataset was not so small, and interactions could be identified, you would see something like below.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>tidySingleCellExperiment<span class="op">::</span>pbmc_small_nested_interactions</span>
<span id="cb28-2"><a href="#cb28-2"></a></span>
<span id="cb28-3"><a href="#cb28-3"></a><span class="co">## # A tibble: 100 x 9</span></span>
<span id="cb28-4"><a href="#cb28-4"></a><span class="co">##    sample ligand receptor ligand.name receptor.name origin destination</span></span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="co">##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;         &lt;chr&gt;  &lt;chr&gt;      </span></span>
<span id="cb28-6"><a href="#cb28-6"></a><span class="co">##  1 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 2  </span></span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="co">##  2 sampl… clust… cluster… B2M         KLRD1         clust… cluster 2  </span></span>
<span id="cb28-8"><a href="#cb28-8"></a><span class="co">##  3 sampl… clust… cluster… IL16        CD4           clust… cluster 2  </span></span>
<span id="cb28-9"><a href="#cb28-9"></a><span class="co">##  4 sampl… clust… cluster… HLA-B       KLRD1         clust… cluster 2  </span></span>
<span id="cb28-10"><a href="#cb28-10"></a><span class="co">##  5 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 2  </span></span>
<span id="cb28-11"><a href="#cb28-11"></a><span class="co">##  6 sampl… clust… cluster… HLA-E       KLRD1         clust… cluster 2  </span></span>
<span id="cb28-12"><a href="#cb28-12"></a><span class="co">##  7 sampl… clust… cluster… GNAS        VIPR1         clust… cluster 2  </span></span>
<span id="cb28-13"><a href="#cb28-13"></a><span class="co">##  8 sampl… clust… cluster… B2M         HFE           clust… cluster 2  </span></span>
<span id="cb28-14"><a href="#cb28-14"></a><span class="co">##  9 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 3  </span></span>
<span id="cb28-15"><a href="#cb28-15"></a><span class="co">## 10 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 3  </span></span>
<span id="cb28-16"><a href="#cb28-16"></a><span class="co">## # … with 90 more rows, and 2 more variables: interaction.type &lt;chr&gt;,</span></span>
<span id="cb28-17"><a href="#cb28-17"></a><span class="co">## #   LRscore &lt;dbl&gt;</span></span></code></pre></div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/tidySingleCellExperiment">https://​www.bioconductor.org/​packages/​tidySingleCellExperiment</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidySingleCellExperiment/actions"><img src="https://github.com/stemangiola/tidySingleCellExperiment/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
