<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>'Brings SingleCellExperiment to the Tidyverse  • tidySCE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="'Brings SingleCellExperiment to the Tidyverse ">
<meta property="og:description" content="It creates an invisible layer that allow to see the SingleCellExperiment object 
    as tibble and interact seamlessly with the tidyverse.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidySCE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the tidySCE package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="tidysce---part-of-tidytranscriptomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tidysce---part-of-tidytranscriptomics" class="anchor"></a>tidySCE - part of tidytranscriptomics</h1></div>
<!-- badges: start -->

<p><strong>Brings SingleCellExperiment to the tidyverse!</strong></p>
<p>website: <a href="https://stemangiola.github.io/tidySCE/">stemangiola.github.io/tidySCE/</a></p>
<p>Please have a look also to</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<!---

[![Build Status](https://travis-ci.org/stemangiola/tidySCE.svg?branch=master)](https://travis-ci.org/stemangiola/tidySCE) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/tidySCE/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/tidySCE?branch=master)

--><div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<thead><tr class="header">
<th>SingleCellExperiment-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidySCE</code> is a SingleCellExperiment object, just better</td>
</tr></tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> APIs like for any tibble</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> APIs like for any tibble</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> like for any tibble</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> like for any tibble</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidySCE</code> invisible layer over a SingleCellExperiment object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_transcripts</code></td>
<td>Add transcript-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From CRAN (in submission)</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"tidySCE"</span>)
</pre></div>
<p>From Github</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"stemangiola/tidySCE"</span>)
</pre></div>
</div>
<div id="create-tidysce-the-best-of-both-worlds" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidysce-the-best-of-both-worlds" class="anchor"></a>Create <code>tidySCE</code>, the best of both worlds!</h2>
<p>This is a SingleCellExperiment object but it is evaluated as tibble. So it is fully compatible both with SingleCellExperiment and tidyverse APIs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span> <span class="op">=</span> <span class="kw">tidySCE</span>::<span class="kw"><a href="reference/pbmc_small.html">pbmc_small</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/tidy.html">tidy</a></span>()
</pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span>
</pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">## # A tibble: 80 x 17</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="co">## # … with 70 more rows, and 10 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="co">## #   ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;,</span></span>
<span id="cb5-16"><a href="#cb5-16"></a><span class="co">## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p><strong>But it is a SingleCellExperiment object after all</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span><span class="op">@</span>assays
</pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co">## An object of class "SimpleAssays"</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">## Slot "data":</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">## List of length 2</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">## names(2): counts logcounts</span></span></code></pre></div>
</div>
<div id="annotation-polishing-using-tidyverse" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation-polishing-using-tidyverse" class="anchor"></a>Annotation polishing using tidyverse</h2>
<p>We may have a column that contains the directory each run was taken from. We may want to extract the run/sample name out of it.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">=</span>
  <span class="kw">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/extract.html">extract</a></span>(<span class="kw">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, remove = <span class="kw">F</span>) 

<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/select.html">select</a></span>(<span class="kw">sample</span>, <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">## # A tibble: 80 x 18</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co">##    cell  sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;  &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        </span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co">##  1 ATGC… sampl… SeuratPro…         70           47 0               A            </span></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co">##  2 CATG… sampl… SeuratPro…         85           52 0               A            </span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co">##  3 GAAC… sampl… SeuratPro…         87           50 1               B            </span></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co">##  4 TGAC… sampl… SeuratPro…        127           56 0               A            </span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="co">##  5 AGTC… sampl… SeuratPro…        173           53 0               A            </span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co">##  6 TCTG… sampl… SeuratPro…         70           48 0               A            </span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co">##  7 TGGT… sampl… SeuratPro…         64           36 0               A            </span></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="co">##  8 GCAG… sampl… SeuratPro…         72           45 0               A            </span></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co">##  9 GATA… sampl… SeuratPro…         52           36 0               A            </span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co">## 10 AATG… sampl… SeuratPro…        100           41 0               A            </span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="co">## # … with 70 more rows, and 11 more variables: groups &lt;chr&gt;,</span></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="co">## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,</span></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="co">## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div id="preliminary-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h2>
<p>We can treat <code>pbmc_small_polished</code> effectively as a normal tibble for plotting.</p>
<p>Here we plot number of transcripts per cell</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="kw">tidySCE</span>::<span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">nFeature_RNA</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span> 
  <span class="fu">geom_histogram</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot1-1.png"><!-- --></p>
<p>Here we plot total transcriptional material per cell</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="kw">tidySCE</span>::<span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">groups</span>, <span class="kw">nCount_RNA</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span> 
  <span class="fu">geom_boxplot</span>(outlier.shape = <span class="fl">NA</span>) <span class="op">+</span>
  <span class="fu">geom_jitter</span>(width = <span class="fl">0.1</span>) <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot2-1.png"><!-- --></p>
<p>Here we plot abundance of two transcripts for each group</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span>(transcripts = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"HLA-DRA"</span> ,     <span class="st">"LYZ"</span> )) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">groups</span>, <span class="kw">abundance_counts</span> <span class="op">+</span> <span class="fl">1</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span> 
  <span class="fu">geom_boxplot</span>(outlier.shape = <span class="fl">NA</span>) <span class="op">+</span> 
  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(size=<span class="kw">nCount_RNA</span>), alpha=<span class="fl">0.5</span>, width = <span class="fl">0.2</span>) <span class="op">+</span> 
  <span class="fu">scale_y_log10</span>() <span class="op">+</span> 
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-9-1.png"><!-- --></p>
</div>
<div id="preprocess-the-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h2>
<p>Also you can treat the object as SingleCellExperiment object and proceed with data processing.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">variable_gens</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
  <span class="fu">modelGeneVar</span>() <span class="op">%&gt;%</span>
  <span class="fu">getTopHVGs</span>(prop=<span class="fl">0.1</span>)

<span class="kw">pbmc_small_pca</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span> 
  <span class="fu">runPCA</span>(   subset_row = <span class="kw">variable_gens</span>  )

<span class="kw">pbmc_small_pca</span>
</pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co">## # A tibble: 80 x 18</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">##  1 ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">##  2 CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="co">##  3 GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="co">##  4 TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">##  5 AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">##  6 TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">##  7 TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="co">##  8 GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co">##  9 GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co">## 10 AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="co">## # … with 70 more rows, and 11 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,</span></span>
<span id="cb14-15"><a href="#cb14-15"></a><span class="co">## #   sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;,</span></span>
<span id="cb14-16"><a href="#cb14-16"></a><span class="co">## #   PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>If a tool is not included in the tidySCE collection, we can use <code>as_tibble</code> to permanently convert <code>tidySCE</code> into tibble</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="kw">as_tibble</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/select.html">select</a></span>(<span class="fu">contains</span>(<span class="st">"PC"</span>), <span class="fu">everything</span>()) <span class="op">%&gt;%</span>
      <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns = <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">groups</span>))
</pre></div>
<p><img src="reference/figures/pc_plot-1.png"><!-- --></p>
</div>
<div id="identify-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h2>
<p>We proceed with cluster identification</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">pbmc_small_cluster</span> <span class="op">=</span> <span class="kw">pbmc_small_pca</span>


<span class="co"># Assigning to the 'colLabels' of the 'sce'.</span>
<span class="fu">colLabels</span>(<span class="kw">pbmc_small_cluster</span>) <span class="op">=</span>
  <span class="kw">pbmc_small_pca</span> <span class="op">%&gt;%</span>
  <span class="fu">buildSNNGraph</span>(use.dimred=<span class="st">"PCA"</span>) <span class="op">%&gt;%</span>
  <span class="kw">igraph</span>::<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>() <span class="op">%$%</span>
  <span class="kw">membership</span> <span class="op">%&gt;%</span>
  <span class="kw">as.factor</span>

<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/select.html">select</a></span>(<span class="kw">label</span>, <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co">## # A tibble: 80 x 19</span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="co">##    cell  label orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="co">##    &lt;chr&gt; &lt;fct&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        </span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">##  1 ATGC… 2     SeuratPro…         70           47 0               A            </span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">##  2 CATG… 2     SeuratPro…         85           52 0               A            </span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">##  3 GAAC… 2     SeuratPro…         87           50 1               B            </span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">##  4 TGAC… 1     SeuratPro…        127           56 0               A            </span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">##  5 AGTC… 2     SeuratPro…        173           53 0               A            </span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co">##  6 TCTG… 2     SeuratPro…         70           48 0               A            </span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co">##  7 TGGT… 1     SeuratPro…         64           36 0               A            </span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co">##  8 GCAG… 2     SeuratPro…         72           45 0               A            </span></span>
<span id="cb17-12"><a href="#cb17-12"></a><span class="co">##  9 GATA… 2     SeuratPro…         52           36 0               A            </span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co">## 10 AATG… 2     SeuratPro…        100           41 0               A            </span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="co">## # … with 70 more rows, and 12 more variables: groups &lt;chr&gt;,</span></span>
<span id="cb17-15"><a href="#cb17-15"></a><span class="co">## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;,</span></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="co">## #   PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</span></span></code></pre></div>
<p>Now we can interrogate the object as if it was a regular tibble data frame</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="kw">tidySCE</span>::<span class="fu"><a href="reference/count.html">count</a></span>(<span class="kw">groups</span>, <span class="kw">label</span>)
</pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co">## # A tibble: 8 x 3</span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="co">##   groups label     n</span></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="co">##   &lt;chr&gt;  &lt;fct&gt; &lt;int&gt;</span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co">## 1 g1     1        12</span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co">## 2 g1     2        14</span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">## 3 g1     3        14</span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="co">## 4 g1     4         4</span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co">## 5 g2     1        10</span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">## 6 g2     2        11</span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="co">## 7 g2     3        10</span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="co">## 8 g2     4         5</span></span></code></pre></div>
<p>We can identify cluster markers using SingleCellExperiment</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="co"># Identify markers</span>
<span class="kw">marker_genes</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu">findMarkers</span>(groups=<span class="kw">pbmc_small_cluster</span><span class="op">$</span><span class="kw">label</span>) <span class="op">%&gt;%</span> 
  <span class="kw">as.list</span> <span class="op">%&gt;%</span> 
  <span class="fu">map</span>(<span class="op">~</span> <span class="kw">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">10</span>) <span class="op">%&gt;%</span> <span class="kw">rownames</span>) <span class="op">%&gt;%</span> 
  <span class="kw">unlist</span>

<span class="co"># Plot heatmap</span>
<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span>(transcripts = <span class="kw">marker_genes</span>) <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/group_by.html">group_by</a></span>(<span class="kw">label</span>) <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span>(<span class="kw">transcript</span>, <span class="kw">cell</span>, <span class="kw">abundance_counts</span>, .scale = <span class="st">"column"</span>)
</pre></div>
<p><img src="reference/figures/unnamed-chunk-10-1.png"><!-- --></p>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h2>
<p>We can calculate the first 3 UMAP dimensions using SingleCellExperiment framework</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="kw">pbmc_small_UMAP</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu">runUMAP</span>(ncomponents = <span class="fl">3</span>)
</pre></div>
<p>and we can plot them using 3D plot using plotly</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/plot_ly.html">plot_ly</a></span>(
        x = <span class="op">~</span><span class="kw">`UMAP_1`</span>,
        y = <span class="op">~</span><span class="kw">`UMAP_2`</span>, 
        z = <span class="op">~</span><span class="kw">`UMAP_3`</span>,
        color = <span class="op">~</span> <span class="kw">label</span>
    ) 
</pre></div>
<p><img src="reference/figures/plotly.png" alt="screenshot plotly"></p>
</div>
<div id="cell-type-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h2>
<p>We can infer cell type identities using singleR</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">blueprint</span> <span class="op">=</span> <span class="kw">SingleR</span>::<span class="fu">BlueprintEncodeData</span>()

<span class="kw">cell_type_df</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_UMAP</span><span class="op">@</span>assays<span class="op">@</span>data<span class="op">$</span><span class="kw">logcounts</span> <span class="op">%&gt;%</span>
  <span class="kw">Matrix</span>::<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span>(sparse = <span class="fl">TRUE</span>) <span class="op">%&gt;%</span>
   <span class="kw">SingleR</span>::<span class="fu">SingleR</span>(
       ref = <span class="kw">blueprint</span> ,
       labels = <span class="kw">blueprint</span><span class="op">$</span><span class="kw">label.main</span>,
       method = <span class="st">"single"</span>
   ) <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span>(rownames=<span class="st">"cell"</span>) <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/select.html">select</a></span>(<span class="kw">cell</span>, <span class="kw">first.labels</span>)
</pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">=</span>
  <span class="kw">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/left_join.html">left_join</a></span>(<span class="kw">cell_type_df</span>, by=<span class="st">"cell"</span>)

<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="kw">tidySCE</span>::<span class="fu"><a href="reference/select.html">select</a></span>(<span class="kw">cell</span>, <span class="kw">first.labels</span>, <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co">## # A tibble: 80 x 23</span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="co">##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          </span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              </span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="co">##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              </span></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="co">##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              </span></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="co">##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              </span></span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="co">##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              </span></span>
<span id="cb25-9"><a href="#cb25-9"></a><span class="co">##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              </span></span>
<span id="cb25-10"><a href="#cb25-10"></a><span class="co">##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              </span></span>
<span id="cb25-11"><a href="#cb25-11"></a><span class="co">##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              </span></span>
<span id="cb25-12"><a href="#cb25-12"></a><span class="co">##  9 GATA… CD4+ T-cells SeuratPro…         52           36 0              </span></span>
<span id="cb25-13"><a href="#cb25-13"></a><span class="co">## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              </span></span>
<span id="cb25-14"><a href="#cb25-14"></a><span class="co">## # … with 70 more rows, and 17 more variables: letter.idents &lt;fct&gt;,</span></span>
<span id="cb25-15"><a href="#cb25-15"></a><span class="co">## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;,</span></span>
<span id="cb25-16"><a href="#cb25-16"></a><span class="co">## #   label &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;,</span></span>
<span id="cb25-17"><a href="#cb25-17"></a><span class="co">## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;</span></span></code></pre></div>
<p>We can easily summarise the results. For example, see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/count.html">count</a></span>(<span class="kw">label</span>, <span class="kw">first.labels</span>)
</pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co">## # A tibble: 11 x 3</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="co">##    label first.labels     n</span></span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="co">##    &lt;fct&gt; &lt;chr&gt;        &lt;int&gt;</span></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co">##  1 1     CD4+ T-cells     2</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="co">##  2 1     CD8+ T-cells     8</span></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="co">##  3 1     NK cells        12</span></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="co">##  4 2     B-cells         10</span></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="co">##  5 2     CD4+ T-cells     6</span></span>
<span id="cb27-9"><a href="#cb27-9"></a><span class="co">##  6 2     CD8+ T-cells     2</span></span>
<span id="cb27-10"><a href="#cb27-10"></a><span class="co">##  7 2     Macrophages      1</span></span>
<span id="cb27-11"><a href="#cb27-11"></a><span class="co">##  8 2     Monocytes        6</span></span>
<span id="cb27-12"><a href="#cb27-12"></a><span class="co">##  9 3     Macrophages      1</span></span>
<span id="cb27-13"><a href="#cb27-13"></a><span class="co">## 10 3     Monocytes       23</span></span>
<span id="cb27-14"><a href="#cb27-14"></a><span class="co">## 11 4     Erythrocytes     9</span></span></code></pre></div>
<p>We can easily reshape the data for building information-rish faceted plots</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Reshaping</span>
  <span class="fu"><a href="reference/pivot_longer.html">pivot_longer</a></span>(
    cols=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">label</span>, <span class="kw">first.labels</span>), 
    names_to = <span class="st">"classifier"</span>, values_to = <span class="st">"label"</span>
  ) <span class="op">%&gt;%</span>
  
  <span class="co"># Plotting</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">UMAP1</span>, <span class="kw">UMAP2</span>, color=<span class="kw">label</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">classifier</span>) <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-14-1.png"><!-- --></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Add mitochondrial abundance</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(mitochondrial = <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fu">n</span>())) <span class="op">%&gt;%</span>
  
  <span class="co"># Plot correlation</span>
  <span class="fu"><a href="reference/join_transcripts.html">join_transcripts</a></span>(transcripts = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CST3"</span> ,     <span class="st">"LYZ"</span> ), shape = <span class="st">"wide"</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">CST3</span> <span class="op">+</span><span class="fl">1</span>, <span class="kw">LYZ</span> <span class="op">+</span> <span class="fl">1</span>, color=<span class="kw">groups</span>, size=<span class="kw">mitochondrial</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span> 
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">first.labels</span>, scales = <span class="st">"free"</span>) <span class="op">+</span>
  <span class="fu">scale_x_log10</span>() <span class="op">+</span>
  <span class="fu">scale_y_log10</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-15-1.png"><!-- --></p>
</div>
<div id="nested-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h2>
<p>A powerful tool we can use with tidySCE is <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="kw">pbmc_small_nested</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/filter.html">filter</a></span>(<span class="kw">first.labels</span> != <span class="st">"Erythrocytes"</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(cell_class = <span class="fu">if_else</span>(<span class="kw">`first.labels`</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span>), <span class="st">"myeloid"</span>, <span class="st">"lmphoid"</span>)) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">nest</a></span>(data = <span class="op">-</span><span class="kw">cell_class</span>)

<span class="kw">pbmc_small_nested</span>
</pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="co">##   cell_class data     </span></span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;   </span></span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="co">## 1 lmphoid    &lt;tidySCE&gt;</span></span>
<span id="cb31-5"><a href="#cb31-5"></a><span class="co">## 2 myeloid    &lt;tidySCE&gt;</span></span></code></pre></div>
<p>Now we can independently (i) find variable features, (ii) reduce dimensions, and (iii) cluster. Using both tidyverse and SingleCellExperiment seamlessy</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_nested</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(data = <span class="fu">map</span>(
    <span class="kw">data</span>, <span class="op">~</span> {
      
      <span class="kw">.x</span> <span class="op">=</span> <span class="fu">runPCA</span>(<span class="kw">.x</span>, subset_row = <span class="kw">variable_gens</span> )
      
      <span class="kw">variable_gens</span> <span class="op">=</span> 
        <span class="kw">.x</span> <span class="op">%&gt;%</span>
        <span class="fu">modelGeneVar</span>() <span class="op">%&gt;%</span>
        <span class="fu">getTopHVGs</span>(prop=<span class="fl">0.3</span>)
      
      <span class="fu">colLabels</span>(<span class="kw">.x</span>) <span class="op">=</span>
        <span class="kw">.x</span> <span class="op">%&gt;%</span>
        <span class="fu">buildSNNGraph</span>(use.dimred=<span class="st">"PCA"</span>) <span class="op">%&gt;%</span>
        <span class="kw">igraph</span>::<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>() <span class="op">%$%</span>
        <span class="kw">membership</span> <span class="op">%&gt;%</span>
        <span class="kw">as.factor</span>
      
      <span class="kw">.x</span> <span class="op">%&gt;%</span> <span class="fu">runUMAP</span>(ncomponents = <span class="fl">3</span>)
      
    }
  )) 

<span class="kw">pbmc_small_nested_reanalysed</span>
</pre></div>
<div class="sourceCode" id="cb33"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="co">## # A tibble: 2 x 2</span></span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="co">##   cell_class data     </span></span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="co">##   &lt;chr&gt;      &lt;list&gt;   </span></span>
<span id="cb33-4"><a href="#cb33-4"></a><span class="co">## 1 lmphoid    &lt;tidySCE&gt;</span></span>
<span id="cb33-5"><a href="#cb33-5"></a><span class="co">## 2 myeloid    &lt;tidySCE&gt;</span></span></code></pre></div>
<p>Now we can unnest and plot the new clasification</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Convert to tibble otherwise SingleCellExperiment drops reduced dimensions when unifying data sets.</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(data = <span class="fu">map</span>(<span class="kw">data</span>, <span class="op">~</span> <span class="kw">.x</span> <span class="op">%&gt;%</span> <span class="kw">as_tibble</span>)) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">unnest</a></span>(<span class="kw">data</span>) <span class="op">%&gt;%</span>

  <span class="co"># Define unique clusters</span>
  <span class="fu"><a href="reference/unite.html">unite</a></span>(<span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">cell_class</span>, <span class="kw">label</span>), remove=<span class="fl">FALSE</span>) <span class="op">%&gt;%</span>
  
  <span class="co"># Plotting</span>
  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span>(<span class="fu">aes</span>(<span class="kw">UMAP1</span>, <span class="kw">UMAP2</span>, color=<span class="kw">cluster</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">cell_class</span>) <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-18-1.png"><!-- --></p>
<p>We can perform a large number of functional analyses on data subsets. For example intra-sample cell-cell interactions, for then comparing/testing whether interactions have strenghten/weaken across conditions. In this small dataset we have just two sample (one for condition) but you can imagine how you can use tidyverse to perform t-tests and visualisation.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="kw">pbmc_small_nested_interactions</span> <span class="op">=</span> 
  <span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Unnest based on cell category</span>
  <span class="fu"><a href="reference/unnest.html">unnest</a></span>(<span class="kw">data</span>) <span class="op">%&gt;%</span>
  
  <span class="co"># Create unambiguous clusters</span>
  <span class="fu"><a href="reference/mutate.html">mutate</a></span>(integrated_clusters = <span class="kw">first.labels</span> <span class="op">%&gt;%</span> <span class="kw">as.factor</span> <span class="op">%&gt;%</span> <span class="kw">as.integer</span>) <span class="op">%&gt;%</span>

  <span class="co"># Nest based on sample</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="reference/unnest.html">nest</a></span>(data = <span class="op">-</span><span class="kw">sample</span>) <span class="op">%&gt;%</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="reference/mutate.html">mutate</a></span>(interactions = <span class="fu">map</span>(<span class="kw">data</span>, <span class="op">~</span> {
        
        <span class="co"># Produce variables. Yuck!</span>
        <span class="kw">cluster</span> <span class="op">=</span> <span class="kw">.x</span><span class="op">@</span>colData<span class="op">$</span><span class="kw">integrated_clusters</span>
        <span class="kw">data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">.x</span><span class="op">@</span>assays<span class="op">@</span>data <span class="op">%&gt;%</span> <span class="kw">as.list</span> <span class="op">%&gt;%</span> <span class="kw">.</span>[[<span class="fl">1</span>]] <span class="op">%&gt;%</span> <span class="kw">as.matrix</span>)

        <span class="co"># Ligand/Receptor analysis using SingleCellSignalR</span>
        <span class="kw">data</span> <span class="op">%&gt;%</span>
            <span class="fu">cell_signaling</span>(genes=<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">data</span>),cluster=<span class="kw">cluster</span>) <span class="op">%&gt;%</span>
            <span class="fu">inter_network</span>(data = <span class="kw">data</span>, signal = <span class="kw">.</span>, genes = <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">data</span>), cluster = <span class="kw">cluster</span>) <span class="op">%$%</span>
            <span class="kw">`individual-networks`</span> <span class="op">%&gt;%</span> 
            <span class="fu">map_dfr</span>(<span class="op">~</span><span class="fu"><a href="reference/arrange.html">bind_rows</a></span>(<span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span>(<span class="kw">.x</span>))) 
    })) 

<span class="kw">pbmc_small_nested_interactions</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/select.html">select</a></span>(<span class="op">-</span><span class="kw">data</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/unnest.html">unnest</a></span>(<span class="kw">interactions</span>)
</pre></div>
<p>If the data set was not so small, and interactions could be identified, you would see something as below.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="co">## # A tibble: 100 x 9</span></span>
<span id="cb36-2"><a href="#cb36-2"></a><span class="co">##    sample ligand receptor ligand.name receptor.name origin destination</span></span>
<span id="cb36-3"><a href="#cb36-3"></a><span class="co">##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;         &lt;chr&gt;  &lt;chr&gt;      </span></span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="co">##  1 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 2  </span></span>
<span id="cb36-5"><a href="#cb36-5"></a><span class="co">##  2 sampl… clust… cluster… B2M         KLRD1         clust… cluster 2  </span></span>
<span id="cb36-6"><a href="#cb36-6"></a><span class="co">##  3 sampl… clust… cluster… IL16        CD4           clust… cluster 2  </span></span>
<span id="cb36-7"><a href="#cb36-7"></a><span class="co">##  4 sampl… clust… cluster… HLA-B       KLRD1         clust… cluster 2  </span></span>
<span id="cb36-8"><a href="#cb36-8"></a><span class="co">##  5 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 2  </span></span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="co">##  6 sampl… clust… cluster… HLA-E       KLRD1         clust… cluster 2  </span></span>
<span id="cb36-10"><a href="#cb36-10"></a><span class="co">##  7 sampl… clust… cluster… GNAS        VIPR1         clust… cluster 2  </span></span>
<span id="cb36-11"><a href="#cb36-11"></a><span class="co">##  8 sampl… clust… cluster… B2M         HFE           clust… cluster 2  </span></span>
<span id="cb36-12"><a href="#cb36-12"></a><span class="co">##  9 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 3  </span></span>
<span id="cb36-13"><a href="#cb36-13"></a><span class="co">## 10 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 3  </span></span>
<span id="cb36-14"><a href="#cb36-14"></a><span class="co">## # … with 90 more rows, and 2 more variables: interaction.type &lt;chr&gt;,</span></span>
<span id="cb36-15"><a href="#cb36-15"></a><span class="co">## #   LRscore &lt;dbl&gt;</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidySCE/actions"><img src="https://github.com/stemangiola/tidySCE/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
