<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the tidySCE package • tidySCE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the tidySCE package">
<meta property="og:description" content="tidySCE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidySCE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the tidySCE package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="introduction_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the tidySCE package</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            
            <h4 class="date">2020-09-07</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>tidySCE provides a bridge between Bioconductor single-cell packages <span class="citation">(Amezquita et al. 2019)</span> and the tidyverse <span class="citation">(Wickham et al. 2019)</span>. It creates an invisible layer that enables viewing the Bioconductor <em>SingleCellExperiment</em> object as a tidyverse tibble, and provides SingleCellExperiment-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions. This allows users to get the best of both Bioconductor and tidyverse worlds.</p>
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<thead><tr class="header">
<th>SingleCellExperiment-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidySCE</code> is a SingleCellExperiment object, just better</td>
</tr></tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> tibble functions (e.g. <code><a href="../reference/select.html">tidySCE::select</a></code>)</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> tibble functions (e.g. <code><a href="../reference/pivot_longer.html">tidySCE::pivot_longer</a></code>)</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> (<code><a href="../reference/ggplot.html">tidySCE::ggplot</a></code>)</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> (<code><a href="../reference/plot_ly.html">tidySCE::plot_ly</a></code>)</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidySCE</code> invisible layer over a SingleCellExperiment object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_transcripts</code></td>
<td>Add transcript-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, quietly=<span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)

<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"tidySCE"</span>)</pre></div>
<p>Load libraries used in this vignette.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Bioconductor single-cell packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioconductor.org/packages/scater/">scater</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">scran</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/LTLA/SingleR">SingleR</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">SingleCellSignalR</span>)

<span class="co"># Tidyverse-compatible packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://purrr.tidyverse.org">purrr</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://www.r-project.org">tidyHeatmap</a></span>)

<span class="co"># Both</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">tidySCE</span>)</pre></div>
</div>
</div>
<div id="create-tidysce-the-best-of-both-worlds" class="section level1">
<h1 class="hasAnchor">
<a href="#create-tidysce-the-best-of-both-worlds" class="anchor"></a>Create <code>tidySCE</code>, the best of both worlds!</h1>
<p>This is a <em>SingleCellExperiment</em> object but it is evaluated as a tibble. So it is compatible both with SingleCellExperiment and tidyverse.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span> <span class="op">&lt;-</span> <span class="kw">tidySCE</span>::<span class="kw"><a href="../reference/pbmc_small.html">pbmc_small</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/tidy.html">tidy</a></span>()</pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span></pre></div>
<pre><code>## # A tibble: 80 x 17
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 ATGC… SeuratPro…         70           47 0               A             g2    
##  2 CATG… SeuratPro…         85           52 0               A             g1    
##  3 GAAC… SeuratPro…         87           50 1               B             g2    
##  4 TGAC… SeuratPro…        127           56 0               A             g2    
##  5 AGTC… SeuratPro…        173           53 0               A             g2    
##  6 TCTG… SeuratPro…         70           48 0               A             g1    
##  7 TGGT… SeuratPro…         64           36 0               A             g1    
##  8 GCAG… SeuratPro…         72           45 0               A             g1    
##  9 GATA… SeuratPro…         52           36 0               A             g1    
## 10 AATG… SeuratPro…        100           41 0               A             g1    
## # … with 70 more rows, and 10 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;,
## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
<p><strong>But it is a SingleCellExperiment object after all</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span><span class="op">@</span>assays</pre></div>
<pre><code>## An object of class "SimpleAssays"
## Slot "data":
## List of length 2
## names(2): counts logcounts</code></pre>
</div>
<div id="annotation-polishing" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-polishing" class="anchor"></a>Annotation polishing</h1>
<p>We may have a column that contains the directory each run was taken from, such as the “file” column in <code>pbmc_small_tidy</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">pbmc_small_tidy</span><span class="op">$</span><span class="kw">file</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">5</span>]</pre></div>
<pre><code>## [1] "../data/sample2/outs/filtered_feature_bc_matrix/"
## [2] "../data/sample1/outs/filtered_feature_bc_matrix/"
## [3] "../data/sample2/outs/filtered_feature_bc_matrix/"
## [4] "../data/sample2/outs/filtered_feature_bc_matrix/"
## [5] "../data/sample2/outs/filtered_feature_bc_matrix/"</code></pre>
<p>We may want to extract the run/sample name out of it into a separate column. Tidyverse <code>extract</code> can be used to convert a character column into multiple columns using regular expression groups.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># Create sample column</span>
<span class="kw">pbmc_small_polished</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/extract.html">extract</a></span>(<span class="kw">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, remove=<span class="fl">FALSE</span>)

<span class="co"># Reorder to have sample column up front</span>
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/select.html">select</a></span>(<span class="kw">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span>())</pre></div>
<pre><code>## # A tibble: 80 x 18
##    cell  sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents
##    &lt;chr&gt; &lt;chr&gt;  &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        
##  1 ATGC… sampl… SeuratPro…         70           47 0               A            
##  2 CATG… sampl… SeuratPro…         85           52 0               A            
##  3 GAAC… sampl… SeuratPro…         87           50 1               B            
##  4 TGAC… sampl… SeuratPro…        127           56 0               A            
##  5 AGTC… sampl… SeuratPro…        173           53 0               A            
##  6 TCTG… sampl… SeuratPro…         70           48 0               A            
##  7 TGGT… sampl… SeuratPro…         64           36 0               A            
##  8 GCAG… sampl… SeuratPro…         72           45 0               A            
##  9 GATA… sampl… SeuratPro…         52           36 0               A            
## 10 AATG… sampl… SeuratPro…        100           41 0               A            
## # … with 70 more rows, and 11 more variables: groups &lt;chr&gt;,
## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,
## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
</div>
<div id="preliminary-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h1>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="co"># Use colourblind-friendly colours</span>
<span class="kw">friendly_cols</span> <span class="op">&lt;-</span> <span class="kw">dittoSeq</span>::<span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span>()

<span class="co"># Set theme</span>
<span class="kw">my_theme</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(values=<span class="kw">friendly_cols</span>),
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(values=<span class="kw">friendly_cols</span>),
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(
                panel.border=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
                axis.line=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span>(),
                panel.grid.major=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span>(size=<span class="fl">0.2</span>),
                panel.grid.minor=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span>(size=<span class="fl">0.1</span>),
                text=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size=<span class="fl">12</span>),
                legend.position=<span class="st">"bottom"</span>,
                aspect.ratio=<span class="fl">1</span>,
                strip.background=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
                axis.title.x=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(margin=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span>(t=<span class="fl">10</span>, r=<span class="fl">10</span>, b=<span class="fl">10</span>, l=<span class="fl">10</span>)),
                axis.title.y=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(margin=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span>(t=<span class="fl">10</span>, r=<span class="fl">10</span>, b=<span class="fl">10</span>, l=<span class="fl">10</span>))
            )
    )</pre></div>
<p>We can treat <code>pbmc_small_polished</code> as a tibble for plotting.</p>
<p>Here we plot number of transcripts per cell.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="../reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">nFeature_RNA</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/plot1-1.png" width="700"></p>
<p>Here we plot total transcripts per cell.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="../reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">groups</span>, <span class="kw">nCount_RNA</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>(outlier.shape=<span class="fl">NA</span>) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(width=<span class="fl">0.1</span>) <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/plot2-1.png" width="700"></p>
<p>Here we plot abundance of two transcripts for each group.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/join_transcripts.html">join_transcripts</a></span>(transcripts=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span>)) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">groups</span>, <span class="kw">abundance_counts</span> <span class="op">+</span> <span class="fl">1</span>, fill=<span class="kw">groups</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>(outlier.shape=<span class="fl">NA</span>) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(size=<span class="kw">nCount_RNA</span>), alpha=<span class="fl">0.5</span>, width=<span class="fl">0.2</span>) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="preprocess-the-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h1>
<p>We can also treat <code>pbmc_small_polished</code> as a <em>SingleCellExperiment</em> object and proceed with data processing with Bioconductor packages, such as <em>scran</em> <span class="citation">(Lun, Bach, and Marioni 2016)</span> and <em>scater</em> <span class="citation">(McCarthy et al. 2017)</span>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="co"># Identify variable genes with scran</span>
<span class="kw">variable_genes</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span>(prop=<span class="fl">0.1</span>)

<span class="co"># Perform PCA with scater</span>
<span class="kw">pbmc_small_pca</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_polished</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span>(subset_row=<span class="kw">variable_genes</span>)

<span class="kw">pbmc_small_pca</span></pre></div>
<pre><code>## # A tibble: 80 x 18
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 ATGC… SeuratPro…         70           47 0               A             g2    
##  2 CATG… SeuratPro…         85           52 0               A             g1    
##  3 GAAC… SeuratPro…         87           50 1               B             g2    
##  4 TGAC… SeuratPro…        127           56 0               A             g2    
##  5 AGTC… SeuratPro…        173           53 0               A             g2    
##  6 TCTG… SeuratPro…         70           48 0               A             g1    
##  7 TGGT… SeuratPro…         64           36 0               A             g1    
##  8 GCAG… SeuratPro…         72           45 0               A             g1    
##  9 GATA… SeuratPro…         52           36 0               A             g1    
## 10 AATG… SeuratPro…        100           41 0               A             g1    
## # … with 70 more rows, and 11 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;,
## #   PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
<p>If a tidyverse-compatible package is not included in the tidySCE collection, we can use <code>as_tibble</code> to permanently convert <code>tidySCE</code> into a tibble.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="co"># Create pairs plot with GGally</span>
<span class="kw">pbmc_small_pca</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/select.html">select</a></span>(<span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"PC"</span>), <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span>()) <span class="op">%&gt;%</span>
    <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns=<span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">groups</span>)) <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/pc_plot-1.png" width="700"></p>
</div>
<div id="identify-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h1>
<p>We can proceed with cluster identification with <em>scran</em>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="kw">pbmc_small_cluster</span> <span class="op">&lt;-</span> <span class="kw">pbmc_small_pca</span>

<span class="co"># Assign clusters to the 'colLabels' of the SummarizedExperiment object</span>
<span class="fu">colLabels</span>(<span class="kw">pbmc_small_cluster</span>) <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_pca</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(use.dimred=<span class="st">"PCA"</span>) <span class="op">%&gt;%</span>
    <span class="kw">igraph</span>::<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>() <span class="op">%$%</span>
    <span class="kw">membership</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>()

<span class="co"># Reorder columns</span>
<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/select.html">select</a></span>(<span class="kw">label</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span>())</pre></div>
<pre><code>## # A tibble: 80 x 19
##    cell  label orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents
##    &lt;chr&gt; &lt;fct&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        
##  1 ATGC… 2     SeuratPro…         70           47 0               A            
##  2 CATG… 2     SeuratPro…         85           52 0               A            
##  3 GAAC… 2     SeuratPro…         87           50 1               B            
##  4 TGAC… 1     SeuratPro…        127           56 0               A            
##  5 AGTC… 2     SeuratPro…        173           53 0               A            
##  6 TCTG… 2     SeuratPro…         70           48 0               A            
##  7 TGGT… 1     SeuratPro…         64           36 0               A            
##  8 GCAG… 2     SeuratPro…         72           45 0               A            
##  9 GATA… 2     SeuratPro…         52           36 0               A            
## 10 AATG… 2     SeuratPro…        100           41 0               A            
## # … with 70 more rows, and 12 more variables: groups &lt;chr&gt;,
## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;,
## #   PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
<p>And interrogate the output as if it was a regular tibble.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="co"># Count number of cells for each cluster per group</span>
<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="../reference/count.html">count</a></span>(<span class="kw">groups</span>, <span class="kw">label</span>)</pre></div>
<pre><code>## # A tibble: 8 x 3
##   groups label     n
##   &lt;chr&gt;  &lt;fct&gt; &lt;int&gt;
## 1 g1     1        12
## 2 g1     2        14
## 3 g1     3        14
## 4 g1     4         4
## 5 g2     1        10
## 6 g2     2        11
## 7 g2     3        10
## 8 g2     4         5</code></pre>
<p>We can identify and visualise cluster markers combining SingleCellExperiment, tidyverse functions and tidyHeatmap <span class="citation">(Mangiola and Papenfuss 2020)</span></p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="co"># Identify top 10 markers per cluster</span>
<span class="kw">marker_genes</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span>(groups=<span class="kw">pbmc_small_cluster</span><span class="op">$</span><span class="kw">label</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="op">~</span> <span class="kw">.x</span> <span class="op">%&gt;%</span>
        <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">10</span>) <span class="op">%&gt;%</span>
        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>()) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>()

<span class="co"># Plot heatmap</span>
<span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/join_transcripts.html">join_transcripts</a></span>(transcripts=<span class="kw">marker_genes</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/group_by.html">group_by</a></span>(<span class="kw">label</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span>(<span class="kw">transcript</span>, <span class="kw">cell</span>, <span class="kw">abundance_counts</span>, .scale=<span class="st">"column"</span>)</pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div id="reduce-dimensions" class="section level1">
<h1 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h1>
<p>We can calculate the first 3 UMAP dimensions using the SingleCellExperiment framework and <em>scater</em>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="kw">pbmc_small_UMAP</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_cluster</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(ncomponents=<span class="fl">3</span>)</pre></div>
<p>And we can plot the result in 3D using plotly.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="kw">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/plot_ly.html">plot_ly</a></span>(
        x=<span class="op">~</span><span class="kw">`UMAP1`</span>,
        y=<span class="op">~</span><span class="kw">`UMAP2`</span>,
        z=<span class="op">~</span><span class="kw">`UMAP3`</span>,
        color=<span class="op">~</span><span class="kw">label</span>,
        colors=<span class="kw">friendly_cols</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">4</span>]
    )</pre></div>
<p>#<img src="%22../man/figures/plotly.png%22" alt="plotly screenshot"></p>
</div>
<div id="cell-type-prediction" class="section level1">
<h1 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h1>
<p>We can infer cell type identities using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="co"># Get cell type reference data</span>
<span class="kw">blueprint</span> <span class="op">&lt;-</span> <span class="kw">celldex</span>::<span class="fu"><a href="https://rdrr.io/pkg/celldex/man/BlueprintEncodeData.html">BlueprintEncodeData</a></span>()

<span class="co"># Infer cell identities</span>
<span class="kw">cell_type_df</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_UMAP</span><span class="op">@</span>assays<span class="op">@</span>data<span class="op">$</span><span class="kw">logcounts</span> <span class="op">%&gt;%</span>
    <span class="kw">Matrix</span>::<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span>(sparse=<span class="fl">TRUE</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span>(
        ref=<span class="kw">blueprint</span>,
        labels=<span class="kw">blueprint</span><span class="op">$</span><span class="kw">label.main</span>
    ) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span>(rownames=<span class="st">"cell"</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/select.html">select</a></span>(<span class="kw">cell</span>, <span class="kw">first.labels</span>)

<span class="co"># Join UMAP and cell type info</span>
<span class="kw">pbmc_small_cell_type</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_UMAP</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/left_join.html">left_join</a></span>(<span class="kw">cell_type_df</span>, by=<span class="st">"cell"</span>)

<span class="co"># Reorder columns</span>
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="../reference/select.html">select</a></span>(<span class="kw">cell</span>, <span class="kw">first.labels</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span>())</pre></div>
<pre><code>## # A tibble: 80 x 23
##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8
##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          
##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              
##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              
##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              
##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              
##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              
##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              
##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              
##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              
##  9 GATA… CD8+ T-cells SeuratPro…         52           36 0              
## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              
## # … with 70 more rows, and 17 more variables: letter.idents &lt;fct&gt;,
## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;,
## #   label &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;,
## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;</code></pre>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="co"># Count number of cells for each cell type per cluster</span>
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/count.html">count</a></span>(<span class="kw">label</span>, <span class="kw">first.labels</span>)</pre></div>
<pre><code>## # A tibble: 9 x 3
##   label first.labels     n
##   &lt;fct&gt; &lt;chr&gt;        &lt;int&gt;
## 1 1     CD4+ T-cells     2
## 2 1     CD8+ T-cells     8
## 3 1     NK cells        12
## 4 2     B-cells         10
## 5 2     CD4+ T-cells     5
## 6 2     CD8+ T-cells     3
## 7 2     Monocytes        7
## 8 3     Monocytes       24
## 9 4     Erythrocytes     9</code></pre>
<p>We can easily reshape the data for building information-rich faceted plots.</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>

    <span class="co"># Reshape and add classifier column</span>
    <span class="fu"><a href="../reference/pivot_longer.html">pivot_longer</a></span>(
        cols=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">label</span>, <span class="kw">first.labels</span>),
        names_to=<span class="st">"classifier"</span>, values_to=<span class="st">"label"</span>
    ) <span class="op">%&gt;%</span>

    <span class="co"># UMAP plots for cell type and cluster</span>
    <span class="fu"><a href="../reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">UMAP1</span>, <span class="kw">UMAP2</span>, color=<span class="kw">label</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span><span class="kw">classifier</span>) <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>

    <span class="co"># Add some mitochondrial abundance values</span>
    <span class="fu"><a href="../reference/mutate.html">mutate</a></span>(mitochondrial=<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="kw">dplyr</span>::<span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span>())) <span class="op">%&gt;%</span>

    <span class="co"># Plot correlation</span>
    <span class="fu"><a href="../reference/join_transcripts.html">join_transcripts</a></span>(transcripts=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CST3"</span>, <span class="st">"LYZ"</span>), shape=<span class="st">"wide"</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">CST3</span> <span class="op">+</span> <span class="fl">1</span>, <span class="kw">LYZ</span> <span class="op">+</span> <span class="fl">1</span>, color=<span class="kw">groups</span>, size=<span class="kw">mitochondrial</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span><span class="kw">first.labels</span>, scales=<span class="st">"free"</span>) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div id="nested-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h1>
<p>A powerful tool we can use with tidySCE is tidyverse <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types into lymphoid and myeloid, and then nest based on the new classification.</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="kw">pbmc_small_nested</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_cell_type</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/filter.html">filter</a></span>(<span class="kw">first.labels</span> != <span class="st">"Erythrocytes"</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/mutate.html">mutate</a></span>(cell_class=<span class="kw">dplyr</span>::<span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span>(<span class="kw">`first.labels`</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span>), <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span>)) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/unnest.html">nest</a></span>(data=<span class="op">-</span><span class="kw">cell_class</span>)

<span class="kw">pbmc_small_nested</span></pre></div>
<pre><code>## # A tibble: 2 x 2
##   cell_class data     
##   &lt;chr&gt;      &lt;list&gt;   
## 1 lymphoid   &lt;tidySCE&gt;
## 2 myeloid    &lt;tidySCE&gt;</code></pre>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_nested</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/mutate.html">mutate</a></span>(data=<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(
        <span class="kw">data</span>, <span class="op">~</span> {
            <span class="kw">.x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span>(<span class="kw">.x</span>, subset_row=<span class="kw">variable_genes</span>)

            <span class="kw">variable_genes</span> <span class="op">&lt;-</span>
                <span class="kw">.x</span> <span class="op">%&gt;%</span>
                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span>() <span class="op">%&gt;%</span>
                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span>(prop=<span class="fl">0.3</span>)

            <span class="fu">colLabels</span>(<span class="kw">.x</span>) <span class="op">&lt;-</span>
                <span class="kw">.x</span> <span class="op">%&gt;%</span>
                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(use.dimred=<span class="st">"PCA"</span>) <span class="op">%&gt;%</span>
                <span class="kw">igraph</span>::<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>() <span class="op">%$%</span>
                <span class="kw">membership</span> <span class="op">%&gt;%</span>
                <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>()

            <span class="kw">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(ncomponents=<span class="fl">3</span>)
        }
    ))

<span class="kw">pbmc_small_nested_reanalysed</span></pre></div>
<pre><code>## # A tibble: 2 x 2
##   cell_class data     
##   &lt;chr&gt;      &lt;list&gt;   
## 1 lymphoid   &lt;tidySCE&gt;
## 2 myeloid    &lt;tidySCE&gt;</code></pre>
<p>We can then unnest and plot the new classification.</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">%&gt;%</span>

    <span class="co"># Convert to tibble otherwise SingleCellExperiment drops reduced dimensions when unifying data sets.</span>
    <span class="fu"><a href="../reference/mutate.html">mutate</a></span>(data=<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="kw">data</span>, <span class="op">~</span> <span class="kw">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span>())) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/unnest.html">unnest</a></span>(<span class="kw">data</span>) <span class="op">%&gt;%</span>

    <span class="co"># Define unique clusters</span>
    <span class="fu"><a href="../reference/unite.html">unite</a></span>(<span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">cell_class</span>, <span class="kw">label</span>), remove=<span class="fl">FALSE</span>) <span class="op">%&gt;%</span>

    <span class="co"># Plotting</span>
    <span class="fu"><a href="../reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">UMAP1</span>, <span class="kw">UMAP2</span>, color=<span class="kw">cluster</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span><span class="kw">cell_class</span>) <span class="op">+</span>
    <span class="kw">my_theme</span></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>We can perform a large number of functional analyses on data subsets. For example, we can identify intra-sample cell-cell interactions using <em>SingleCellSignalR</em> <span class="citation">(Cabello-Aguilar et al. 2020)</span>, and then compare whether interactions are stronger or weaker across conditions. The code below demonstrates how this analysis could be performed. It won’t work with this small example dataset as we have just two samples (one for each condition). But some example output is shown below and you can imagine how you can use tidyverse on the output to perform t-tests and visualisation.</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="kw">pbmc_small_nested_interactions</span> <span class="op">&lt;-</span>
    <span class="kw">pbmc_small_nested_reanalysed</span> <span class="op">%&gt;%</span>

    <span class="co"># Unnest based on cell category</span>
    <span class="fu"><a href="../reference/unnest.html">unnest</a></span>(<span class="kw">data</span>) <span class="op">%&gt;%</span>

    <span class="co"># Create unambiguous clusters</span>
    <span class="fu"><a href="../reference/mutate.html">mutate</a></span>(integrated_clusters=<span class="kw">first.labels</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>() <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>()) <span class="op">%&gt;%</span>

    <span class="co"># Nest based on sample</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="../reference/unnest.html">nest</a></span>(data=<span class="op">-</span><span class="kw">sample</span>) <span class="op">%&gt;%</span>
    <span class="kw">tidySCE</span>::<span class="fu"><a href="../reference/mutate.html">mutate</a></span>(interactions=<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="kw">data</span>, <span class="op">~</span> {

        <span class="co"># Produce variables. Yuck!</span>
        <span class="kw">cluster</span> <span class="op">&lt;-</span> <span class="kw">.x</span><span class="op">@</span>colData<span class="op">$</span><span class="kw">integrated_clusters</span>
        <span class="kw">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">.x</span><span class="op">@</span>assays<span class="op">@</span>data <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>() <span class="op">%&gt;%</span> <span class="kw">.</span>[[<span class="fl">1</span>]] <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>())

        <span class="co"># Ligand/Receptor analysis using SingleCellSignalR</span>
        <span class="kw">data</span> <span class="op">%&gt;%</span>
            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/cell_signaling.html">cell_signaling</a></span>(genes=<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">data</span>), cluster=<span class="kw">cluster</span>) <span class="op">%&gt;%</span>
            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/inter_network.html">inter_network</a></span>(data=<span class="kw">data</span>, signal=<span class="kw">.</span>, genes=<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">data</span>), cluster=<span class="kw">cluster</span>) <span class="op">%$%</span>
            <span class="kw">`individual-networks`</span> <span class="op">%&gt;%</span>
            <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span>(<span class="op">~</span> <span class="fu"><a href="../reference/arrange.html">bind_rows</a></span>(<span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span>(<span class="kw">.x</span>)))
    }))

<span class="kw">pbmc_small_nested_interactions</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/select.html">select</a></span>(<span class="op">-</span><span class="kw">data</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/unnest.html">unnest</a></span>(<span class="kw">interactions</span>)</pre></div>
<p>If the dataset was not so small, and interactions could be identified, you would see something like below.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="kw">tidySCE</span>::<span class="kw"><a href="../reference/pbmc_small_nested_interactions.html">pbmc_small_nested_interactions</a></span></pre></div>
<pre><code>## # A tibble: 100 x 9
##    sample ligand receptor ligand.name receptor.name origin destination
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;         &lt;chr&gt;  &lt;chr&gt;      
##  1 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 2  
##  2 sampl… clust… cluster… B2M         KLRD1         clust… cluster 2  
##  3 sampl… clust… cluster… IL16        CD4           clust… cluster 2  
##  4 sampl… clust… cluster… HLA-B       KLRD1         clust… cluster 2  
##  5 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 2  
##  6 sampl… clust… cluster… HLA-E       KLRD1         clust… cluster 2  
##  7 sampl… clust… cluster… GNAS        VIPR1         clust… cluster 2  
##  8 sampl… clust… cluster… B2M         HFE           clust… cluster 2  
##  9 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 3  
## 10 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 3  
## # … with 90 more rows, and 2 more variables: interaction.type &lt;chr&gt;,
## #   LRscore &lt;dbl&gt;</code></pre>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></div>
<pre><code>## R version 4.0.2 Patched (2020-09-02 r79126)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-openmp/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] celldex_0.99.1              tidySCE_0.99.0             
##  [3] tidyHeatmap_1.1.4           purrr_0.3.4                
##  [5] SingleCellSignalR_1.1.0     SingleR_1.3.8              
##  [7] scran_1.17.16               scater_1.17.4              
##  [9] ggplot2_3.3.2               SingleCellExperiment_1.11.6
## [11] SummarizedExperiment_1.19.6 DelayedArray_0.15.7        
## [13] matrixStats_0.56.0          Matrix_1.2-18              
## [15] Biobase_2.49.1              GenomicRanges_1.41.6       
## [17] GenomeInfoDb_1.25.11        IRanges_2.23.10            
## [19] S4Vectors_0.27.12           BiocGenerics_0.35.4        
## [21] knitr_1.29                  BiocStyle_2.17.0           
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.1.9               circlize_0.4.10              
##   [3] AnnotationHub_2.21.5          BiocFileCache_1.13.1         
##   [5] systemfonts_0.3.0             plyr_1.8.6                   
##   [7] igraph_1.2.5                  splines_4.0.2                
##   [9] BiocParallel_1.23.2           digest_0.6.25                
##  [11] foreach_1.5.0                 htmltools_0.5.0              
##  [13] viridis_0.5.1                 gdata_2.18.0                 
##  [15] fansi_0.4.1                   magrittr_1.5                 
##  [17] memoise_1.1.0                 cluster_2.1.0                
##  [19] limma_3.45.13                 ComplexHeatmap_2.5.5         
##  [21] pkgdown_1.5.1.9000            colorspace_1.4-1             
##  [23] rappdirs_0.3.1                blob_1.2.1                   
##  [25] ggrepel_0.8.2                 xfun_0.16                    
##  [27] dplyr_1.0.2                   crayon_1.3.4                 
##  [29] RCurl_1.98-1.2                survival_3.2-3               
##  [31] iterators_1.0.12              glue_1.4.2                   
##  [33] gtable_0.3.0                  zlibbioc_1.35.0              
##  [35] XVector_0.29.3                SIMLR_1.15.0                 
##  [37] GetoptLong_1.0.2              BiocSingular_1.5.0           
##  [39] shape_1.4.4                   scales_1.1.1                 
##  [41] pheatmap_1.0.12               DBI_1.1.0                    
##  [43] GGally_2.0.0                  cpp11_0.2.1                  
##  [45] edgeR_3.31.4                  Rcpp_1.0.5                   
##  [47] xtable_1.8-4                  viridisLite_0.3.0            
##  [49] clue_0.3-57                   dqrng_0.2.1                  
##  [51] bit_4.0.4                     rsvd_1.0.3                   
##  [53] httr_1.4.2                    dittoSeq_1.1.7               
##  [55] FNN_1.1.3                     gplots_3.0.4                 
##  [57] RColorBrewer_1.1-2            ellipsis_0.3.1               
##  [59] pkgconfig_2.0.3               reshape_0.8.8                
##  [61] farver_2.0.3                  scuttle_0.99.12              
##  [63] dbplyr_1.4.4                  uwot_0.1.8                   
##  [65] locfit_1.5-9.4                utf8_1.1.4                   
##  [67] later_1.1.0.1                 tidyselect_1.1.0             
##  [69] labeling_0.3                  rlang_0.4.7                  
##  [71] AnnotationDbi_1.51.3          BiocVersion_3.12.0           
##  [73] munsell_0.5.0                 tools_4.0.2                  
##  [75] cli_2.0.2                     ExperimentHub_1.15.3         
##  [77] RSQLite_2.2.0                 generics_0.0.2               
##  [79] ggridges_0.5.2                fastmap_1.0.1                
##  [81] evaluate_0.14                 stringr_1.4.0                
##  [83] yaml_2.2.1                    ragg_0.3.1                   
##  [85] bit64_4.0.5                   fs_1.5.0                     
##  [87] caTools_1.18.0                mime_0.9                     
##  [89] pracma_2.2.9                  compiler_4.0.2               
##  [91] interactiveDisplayBase_1.27.5 curl_4.3                     
##  [93] beeswarm_0.2.3                png_0.1-7                    
##  [95] tibble_3.0.3                  statmod_1.4.34               
##  [97] stringi_1.4.6                 desc_1.2.0                   
##  [99] RSpectra_0.16-0               lattice_0.20-41              
## [101] bluster_0.99.1                multtest_2.45.0              
## [103] vctrs_0.3.4                   pillar_1.4.6                 
## [105] lifecycle_0.2.0               BiocManager_1.30.10          
## [107] GlobalOptions_0.1.2           RcppAnnoy_0.0.16             
## [109] BiocNeighbors_1.7.0           data.table_1.13.0            
## [111] cowplot_1.0.0                 bitops_1.0-6                 
## [113] irlba_2.3.3                   httpuv_1.5.4                 
## [115] R6_2.4.1                      promises_1.1.1               
## [117] bookdown_0.20                 KernSmooth_2.23-17           
## [119] gridExtra_2.3                 vipor_0.4.5                  
## [121] codetools_0.2-16              MASS_7.3-52                  
## [123] gtools_3.8.2                  assertthat_0.2.1             
## [125] rprojroot_1.3-2               rjson_0.2.20                 
## [127] withr_2.2.0                   GenomeInfoDbData_1.2.3       
## [129] grid_4.0.2                    tidyr_1.1.2                  
## [131] rmarkdown_2.3                 DelayedMatrixStats_1.11.1    
## [133] Rtsne_0.15                    shiny_1.5.0                  
## [135] ggbeeswarm_0.6.0</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-amezquita2019orchestrating">
<p>Amezquita, Robert A, Aaron TL Lun, Etienne Becht, Vince J Carey, Lindsay N Carpp, Ludwig Geistlinger, Federico Martini, et al. 2019. “Orchestrating Single-Cell Analysis with Bioconductor.” <em>Nature Methods</em>, 1–9.</p>
</div>
<div id="ref-aran2019reference">
<p>Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, et al. 2019. “Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.” <em>Nature Immunology</em> 20 (2): 163–72.</p>
</div>
<div id="ref-cabello2020singlecellsignalr">
<p>Cabello-Aguilar, Simon, Mélissa Alame, Fabien Kon-Sun-Tack, Caroline Fau, Matthieu Lacroix, and Jacques Colinge. 2020. “SingleCellSignalR: Inference of Intercellular Networks from Single-Cell Transcriptomics.” <em>Nucleic Acids Research</em> 48 (10): e55–e55.</p>
</div>
<div id="ref-lun2016pooling">
<p>Lun, Aaron TL, Karsten Bach, and John C Marioni. 2016. “Pooling Across Cells to Normalize Single-Cell Rna Sequencing Data with Many Zero Counts.” <em>Genome Biology</em> 17 (1): 75.</p>
</div>
<div id="ref-mangiola2020tidyheatmap">
<p>Mangiola, Stefano, and Anthony T Papenfuss. 2020. “TidyHeatmap: An R Package for Modular Heatmap Production Based on Tidy Principles.” <em>Journal of Open Source Software</em> 5 (52): 2472.</p>
</div>
<div id="ref-mccarthy2017scater">
<p>McCarthy, Davis J, Kieran R Campbell, Aaron TL Lun, and Quin F Wills. 2017. “Scater: Pre-Processing, Quality Control, Normalization and Visualization of Single-Cell Rna-Seq Data in R.” <em>Bioinformatics</em> 33 (8): 1179–86.</p>
</div>
<div id="ref-wickham2019welcome">
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the Tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
