<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the tidySingleCellExperiment package • tidySingleCellExperiment</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the tidySingleCellExperiment package">
<meta property="og:description" content="tidySingleCellExperiment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidySingleCellExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the tidySingleCellExperiment package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidySingleCellExperiment/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the tidySingleCellExperiment
package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2022-08-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidySingleCellExperiment/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>tidySingleCellExperiment provides a bridge between Bioconductor
single-cell packages <span class="citation">(Amezquita et al.
2019)</span> and the tidyverse <span class="citation">(Wickham et al.
2019)</span>. It creates an invisible layer that enables viewing the
Bioconductor <em>SingleCellExperiment</em> object as a tidyverse tibble,
and provides SingleCellExperiment-compatible <em>dplyr</em>,
<em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions. This
allows users to get the best of both Bioconductor and tidyverse
worlds.</p>
<div class="section level3">
<h3 id="functionsutilities-available">Functions/utilities available<a class="anchor" aria-label="anchor" href="#functionsutilities-available"></a>
</h3>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>SingleCellExperiment-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidySingleCellExperiment</code> is a
SingleCellExperiment object, just better</td>
</tr></tbody>
</table>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> tibble functions
(e.g. <code><a href="../reference/dplyr-methods.html">tidySingleCellExperiment::select</a></code>)</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> tibble functions
(e.g. <code><a href="../reference/pivot-methods.html">tidySingleCellExperiment::pivot_longer</a></code>)</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code>
(<code><a href="../reference/ggplot2-methods.html">tidySingleCellExperiment::ggplot</a></code>)</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code>
(<code><a href="../reference/plot_ly.html">tidySingleCellExperiment::plot_ly</a></code>)</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidySingleCellExperiment</code> invisible layer over a
SingleCellExperiment object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_features</code></td>
<td>Add feature-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"tidySingleCellExperiment"</span><span class="op">)</span></span></code></pre></div>
<p>Load libraries used in this vignette.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Bioconductor single-cell packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellSignalR</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Tidyverse-compatible packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org" class="external-link">tidyHeatmap</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Both</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="create-tidysinglecellexperiment-the-best-of-both-worlds">Create <code>tidySingleCellExperiment</code>, the best of both
worlds!<a class="anchor" aria-label="anchor" href="#create-tidysinglecellexperiment-the-best-of-both-worlds"></a>
</h2>
<p>This is a <em>SingleCellExperiment</em> object but it is evaluated as
a tibble. So it is compatible both with SingleCellExperiment and
tidyverse.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_tidy</span> <span class="op">&lt;-</span> <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc_small.html">pbmc_small</a></span> </span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_tidy</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SingleCellExperiment-tibble abstraction: 80 × 17</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=230 | Cells=80 | Assays=counts, logcounts</span></span></span>
<span><span class="co">##    .cell      orig.…¹ nCoun…² nFeat…³ RNA_s…⁴ lette…⁵ groups RNA_s…⁶ file  ident</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATGCCAGAA… Seurat…      70      47 0       A       g2     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CATGGCCTG… Seurat…      85      52 0       A       g1     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> GAACCTGAT… Seurat…      87      50 1       B       g2     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TGACTGGAT… Seurat…     127      56 0       A       g2     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AGTCAGACT… Seurat…     173      53 0       A       g2     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCTGATACA… Seurat…      70      48 0       A       g1     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TGGTATCTA… Seurat…      64      36 0       A       g1     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> GCAGCTCTG… Seurat…      72      45 0       A       g1     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> GATATAACA… Seurat…      52      36 0       A       g1     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AATGTTGAC… Seurat…     100      41 0       A       g1     0       ../d… 0    </span></span>
<span><span class="co">## <span style="color: #949494;"># … with 70 more rows, 7 more variables: PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, and abbreviated</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   variable names ¹​orig.ident, ²​nCount_RNA, ³​nFeature_RNA, ⁴​RNA_snn_res.0.8,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ⁵​letter.idents, ⁶​RNA_snn_res.1</span></span></span></code></pre>
<p><strong>But it is a SingleCellExperiment object after
all</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc_small_tidy</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##         ATGCCAGAACGACT CATGGCCTGTGCAT GAACCTGATGAACC TGACTGGATTCTCA</span></span>
<span><span class="co">## MS4A1                .              .              .              .</span></span>
<span><span class="co">## CD79B                1              .              .              .</span></span>
<span><span class="co">## CD79A                .              .              .              .</span></span>
<span><span class="co">## HLA-DRA              .              1              .              .</span></span>
<span><span class="co">## TCL1A                .              .              .              .</span></span>
<span><span class="co">##         AGTCAGACTGCACA</span></span>
<span><span class="co">## MS4A1                .</span></span>
<span><span class="co">## CD79B                .</span></span>
<span><span class="co">## CD79A                .</span></span>
<span><span class="co">## HLA-DRA              1</span></span>
<span><span class="co">## TCL1A                .</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="annotation-polishing">Annotation polishing<a class="anchor" aria-label="anchor" href="#annotation-polishing"></a>
</h2>
<p>We may have a column that contains the directory each run was taken
from, such as the “file” column in <code>pbmc_small_tidy</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_tidy</span><span class="op">$</span><span class="va">file</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span><span class="co">## [2] "../data/sample1/outs/filtered_feature_bc_matrix/"</span></span>
<span><span class="co">## [3] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span><span class="co">## [4] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span><span class="co">## [5] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span></code></pre>
<p>We may want to extract the run/sample name out of it into a separate
column. Tidyverse <code>extract</code> can be used to convert a
character column into multiple columns using regular expression
groups.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create sample column</span></span>
<span><span class="va">pbmc_small_polished</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_tidy</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/extract-methods.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, remove<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Reorder to have sample column up front</span></span>
<span><span class="va">pbmc_small_polished</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SingleCellExperiment-tibble abstraction: 80 × 18</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=230 | Cells=80 | Assays=counts, logcounts</span></span></span>
<span><span class="co">##    .cell     sample orig.…¹ nCoun…² nFeat…³ RNA_s…⁴ lette…⁵ groups RNA_s…⁶ file </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATGCCAGA… sampl… Seurat…      70      47 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CATGGCCT… sampl… Seurat…      85      52 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> GAACCTGA… sampl… Seurat…      87      50 1       B       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TGACTGGA… sampl… Seurat…     127      56 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AGTCAGAC… sampl… Seurat…     173      53 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCTGATAC… sampl… Seurat…      70      48 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TGGTATCT… sampl… Seurat…      64      36 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> GCAGCTCT… sampl… Seurat…      72      45 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> GATATAAC… sampl… Seurat…      52      36 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AATGTTGA… sampl… Seurat…     100      41 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 70 more rows, 8 more variables: ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, and</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   abbreviated variable names ¹​orig.ident, ²​nCount_RNA, ³​nFeature_RNA,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ⁴​RNA_snn_res.0.8, ⁵​letter.idents, ⁶​RNA_snn_res.1</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="preliminary-plots">Preliminary plots<a class="anchor" aria-label="anchor" href="#preliminary-plots"></a>
</h2>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use colourblind-friendly colours</span></span>
<span><span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html" class="external-link">dittoColors</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set theme</span></span>
<span><span class="va">custom_theme</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">friendly_cols</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">friendly_cols</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>                panel.border<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                axis.line<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                panel.grid.major<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>,</span>
<span>                panel.grid.minor<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>,</span>
<span>                text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,</span>
<span>                legend.position<span class="op">=</span><span class="st">"bottom"</span>,</span>
<span>                aspect.ratio<span class="op">=</span><span class="fl">1</span>,</span>
<span>                strip.background<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                axis.title.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>t<span class="op">=</span><span class="fl">10</span>, r<span class="op">=</span><span class="fl">10</span>, b<span class="op">=</span><span class="fl">10</span>, l<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                axis.title.y<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>t<span class="op">=</span><span class="fl">10</span>, r<span class="op">=</span><span class="fl">10</span>, b<span class="op">=</span><span class="fl">10</span>, l<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>We can treat <code>pbmc_small_polished</code> as a tibble for
plotting.</p>
<p>Here we plot number of features per cell.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_polished</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">nFeature_RNA</span>, fill<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot1-1.png" width="700"></p>
<p>Here we plot total features per cell.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_polished</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">nCount_RNA</span>, fill<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot2-1.png" width="700"></p>
<p>Here we plot abundance of two features for each group.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_polished</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/join_features.html">join_features</a></span><span class="op">(</span>features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">.abundance_counts</span> <span class="op">+</span> <span class="fl">1</span>, fill<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">nCount_RNA</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, width<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="preprocess-the-dataset">Preprocess the dataset<a class="anchor" aria-label="anchor" href="#preprocess-the-dataset"></a>
</h2>
<p>We can also treat <code>pbmc_small_polished</code> as a
<em>SingleCellExperiment</em> object and proceed with data processing
with Bioconductor packages, such as <em>scran</em> <span class="citation">(Lun, Bach, and Marioni 2016)</span> and
<em>scater</em> <span class="citation">(McCarthy et al.
2017)</span>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify variable genes with scran</span></span>
<span><span class="va">variable_genes</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_polished</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span>prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform PCA with scater</span></span>
<span><span class="va">pbmc_small_pca</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_polished</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span>subset_row<span class="op">=</span><span class="va">variable_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_small_pca</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SingleCellExperiment-tibble abstraction: 80 × 18</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=230 | Cells=80 | Assays=counts, logcounts</span></span></span>
<span><span class="co">##    .cell     orig.…¹ nCoun…² nFeat…³ RNA_s…⁴ lette…⁵ groups RNA_s…⁶ file  sample</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATGCCAGA… Seurat…      70      47 0       A       g2     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CATGGCCT… Seurat…      85      52 0       A       g1     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> GAACCTGA… Seurat…      87      50 1       B       g2     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TGACTGGA… Seurat…     127      56 0       A       g2     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AGTCAGAC… Seurat…     173      53 0       A       g2     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCTGATAC… Seurat…      70      48 0       A       g1     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TGGTATCT… Seurat…      64      36 0       A       g1     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> GCAGCTCT… Seurat…      72      45 0       A       g1     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> GATATAAC… Seurat…      52      36 0       A       g1     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AATGTTGA… Seurat…     100      41 0       A       g1     0       ../d… sampl…</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 70 more rows, 8 more variables: ident &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, and</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   abbreviated variable names ¹​orig.ident, ²​nCount_RNA, ³​nFeature_RNA,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ⁴​RNA_snn_res.0.8, ⁵​letter.idents, ⁶​RNA_snn_res.1</span></span></span></code></pre>
<p>If a tidyverse-compatible package is not included in the
tidySingleCellExperiment collection, we can use <code>as_tibble</code>
to permanently convert <code>tidySingleCellExperiment</code> into a
tibble.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create pairs plot with GGally</span></span>
<span><span class="va">pbmc_small_pca</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"PC"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/pc_plot-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="identify-clusters">Identify clusters<a class="anchor" aria-label="anchor" href="#identify-clusters"></a>
</h2>
<p>We can proceed with cluster identification with <em>scran</em>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_cluster</span> <span class="op">&lt;-</span> <span class="va">pbmc_small_pca</span></span>
<span></span>
<span><span class="co"># Assign clusters to the 'colLabels' of the SingleCellExperiment object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">pbmc_small_cluster</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_pca</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span>use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%$%</span></span>
<span>    <span class="va">membership</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Reorder columns</span></span>
<span><span class="va">pbmc_small_cluster</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">label</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SingleCellExperiment-tibble abstraction: 80 × 19</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=230 | Cells=80 | Assays=counts, logcounts</span></span></span>
<span><span class="co">##    .cell      label orig.…¹ nCoun…² nFeat…³ RNA_s…⁴ lette…⁵ groups RNA_s…⁶ file </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATGCCAGAA… 2     Seurat…      70      47 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CATGGCCTG… 2     Seurat…      85      52 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> GAACCTGAT… 2     Seurat…      87      50 1       B       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TGACTGGAT… 1     Seurat…     127      56 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AGTCAGACT… 2     Seurat…     173      53 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCTGATACA… 2     Seurat…      70      48 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TGGTATCTA… 1     Seurat…      64      36 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> GCAGCTCTG… 2     Seurat…      72      45 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> GATATAACA… 2     Seurat…      52      36 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AATGTTGAC… 2     Seurat…     100      41 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 70 more rows, 9 more variables: sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, and</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   abbreviated variable names ¹​orig.ident, ²​nCount_RNA, ³​nFeature_RNA,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ⁴​RNA_snn_res.0.8, ⁵​letter.idents, ⁶​RNA_snn_res.1</span></span></span></code></pre>
<p>And interrogate the output as if it was a regular tibble.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Count number of cells for each cluster per group</span></span>
<span><span class="va">pbmc_small_cluster</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/count.html">count</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">label</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8 × 3</span></span></span>
<span><span class="co">##   groups label     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> g1     1        12</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> g1     2        14</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> g1     3        14</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> g1     4         4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> g2     1        10</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> g2     2        11</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> g2     3        10</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> g2     4         5</span></span></code></pre>
<p>We can identify and visualise cluster markers combining
SingleCellExperiment, tidyverse functions and tidyHeatmap <span class="citation">(Mangiola and Papenfuss 2020)</span></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify top 10 markers per cluster</span></span>
<span><span class="va">marker_genes</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_cluster</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span>groups<span class="op">=</span><span class="va">pbmc_small_cluster</span><span class="op">$</span><span class="va">label</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="op">~</span> <span class="va">.x</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot heatmap</span></span>
<span><span class="va">pbmc_small_cluster</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/join_features.html">join_features</a></span><span class="op">(</span>features<span class="op">=</span><span class="va">marker_genes</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">group_by</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-method.html" class="external-link">heatmap</a></span><span class="op">(</span><span class="va">.feature</span>, <span class="va">.cell</span>, <span class="va">.abundance_counts</span>, .scale<span class="op">=</span><span class="st">"column"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="reduce-dimensions">Reduce dimensions<a class="anchor" aria-label="anchor" href="#reduce-dimensions"></a>
</h2>
<p>We can calculate the first 3 UMAP dimensions using the
SingleCellExperiment framework and <em>scater</em>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_UMAP</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_cluster</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span>ncomponents<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>And we can plot the result in 3D using plotly.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_UMAP</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/plot_ly.html">plot_ly</a></span><span class="op">(</span></span>
<span>        x<span class="op">=</span><span class="op">~</span><span class="va">`UMAP1`</span>,</span>
<span>        y<span class="op">=</span><span class="op">~</span><span class="va">`UMAP2`</span>,</span>
<span>        z<span class="op">=</span><span class="op">~</span><span class="va">`UMAP3`</span>,</span>
<span>        color<span class="op">=</span><span class="op">~</span><span class="va">label</span>,</span>
<span>        colors<span class="op">=</span><span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../reference/figures/plotly.png" alt=""><p class="caption">plotly screenshot</p>
</div>
</div>
<div class="section level2">
<h2 id="cell-type-prediction">Cell type prediction<a class="anchor" aria-label="anchor" href="#cell-type-prediction"></a>
</h2>
<p>We can infer cell type identities using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output
using tidyverse.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get cell type reference data</span></span>
<span><span class="va">blueprint</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/BlueprintEncodeData.html" class="external-link">BlueprintEncodeData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Infer cell identities</span></span>
<span><span class="va">cell_type_df</span> <span class="op">&lt;-</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">pbmc_small_UMAP</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span>sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">SingleR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span></span>
<span>        ref <span class="op">=</span> <span class="va">blueprint</span>,</span>
<span>        labels <span class="op">=</span> <span class="va">blueprint</span><span class="op">$</span><span class="va">label.main</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"single"</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span><span class="op">(</span>rownames<span class="op">=</span><span class="st">"cell"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Join UMAP and cell type info</span></span>
<span><span class="va">pbmc_small_cell_type</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_UMAP</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">left_join</a></span><span class="op">(</span><span class="va">cell_type_df</span>, by<span class="op">=</span><span class="st">"cell"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Reorder columns</span></span>
<span><span class="va">pbmc_small_cell_type</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SingleCellExperiment-tibble abstraction: 80 × 23</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=230 | Cells=80 | Assays=counts, logcounts</span></span></span>
<span><span class="co">##    cell     first…¹ orig.…² nCoun…³ nFeat…⁴ RNA_s…⁵ lette…⁶ groups RNA_s…⁷ file </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATGCCAG… CD4+ T… Seurat…      70      47 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CATGGCC… CD8+ T… Seurat…      85      52 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> GAACCTG… CD8+ T… Seurat…      87      50 1       B       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TGACTGG… CD4+ T… Seurat…     127      56 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AGTCAGA… CD4+ T… Seurat…     173      53 0       A       g2     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TCTGATA… CD4+ T… Seurat…      70      48 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TGGTATC… CD4+ T… Seurat…      64      36 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> GCAGCTC… CD4+ T… Seurat…      72      45 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> GATATAA… CD4+ T… Seurat…      52      36 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AATGTTG… CD4+ T… Seurat…     100      41 0       A       g1     0       ../d…</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 70 more rows, 13 more variables: sample &lt;chr&gt;, ident &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   label &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;, and</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   abbreviated variable names ¹​first.labels, ²​orig.ident, ³​nCount_RNA,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ⁴​nFeature_RNA, ⁵​RNA_snn_res.0.8, ⁶​letter.idents, ⁷​RNA_snn_res.1</span></span></span></code></pre>
<p>We can easily summarise the results. For example, we can see how cell
type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Count number of cells for each cell type per cluster</span></span>
<span><span class="va">pbmc_small_cell_type</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/count.html">count</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 11 × 3</span></span></span>
<span><span class="co">##    label first.labels     n</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 1     CD4+ T-cells     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 1     CD8+ T-cells     8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 1     NK cells        12</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 2     B-cells         10</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 2     CD4+ T-cells     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 2     CD8+ T-cells     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 2     Macrophages      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 2     Monocytes        6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 3     Macrophages      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 3     Monocytes       23</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> 4     Erythrocytes     9</span></span></code></pre>
<p>We can easily reshape the data for building information-rich faceted
plots.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_cell_type</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Reshape and add classifier column</span></span>
<span>    <span class="fu"><a href="../reference/pivot-methods.html">pivot_longer</a></span><span class="op">(</span></span>
<span>        cols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span>,</span>
<span>        names_to<span class="op">=</span><span class="st">"classifier"</span>, values_to<span class="op">=</span><span class="st">"label"</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># UMAP plots for cell type and cluster</span></span>
<span>    <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>, <span class="va">UMAP2</span>, color<span class="op">=</span><span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">classifier</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We can easily plot gene correlation per cell category, adding
multi-layer annotations.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_cell_type</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Add some mitochondrial abundance values</span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span><span class="op">(</span>mitochondrial<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Plot correlation</span></span>
<span>    <span class="fu"><a href="../reference/join_features.html">join_features</a></span><span class="op">(</span>features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CST3"</span>, <span class="st">"LYZ"</span><span class="op">)</span>, shape<span class="op">=</span><span class="st">"wide"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">CST3</span> <span class="op">+</span> <span class="fl">1</span>, <span class="va">LYZ</span> <span class="op">+</span> <span class="fl">1</span>, color<span class="op">=</span><span class="va">groups</span>, size<span class="op">=</span><span class="va">mitochondrial</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">first.labels</span>, scales<span class="op">=</span><span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="nested-analyses">Nested analyses<a class="anchor" aria-label="anchor" href="#nested-analyses"></a>
</h2>
<p>A powerful tool we can use with tidySingleCellExperiment is tidyverse
<code>nest</code>. We can easily perform independent analyses on subsets
of the dataset. First we classify cell types into lymphoid and myeloid,
and then nest based on the new classification.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_cell_type</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">first.labels</span> <span class="op">!=</span> <span class="st">"Erythrocytes"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span><span class="op">(</span>cell_class<span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">`first.labels`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span><span class="op">)</span>, <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/nest-methods.html">nest</a></span><span class="op">(</span>data<span class="op">=</span><span class="op">-</span><span class="va">cell_class</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_small_nested</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   cell_class data           </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> lymphoid   <span style="color: #949494;">&lt;SnglCllE[,40]&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> myeloid    <span style="color: #949494;">&lt;SnglCllE[,31]&gt;</span></span></span></code></pre>
<p>Now we can independently for the lymphoid and myeloid subsets (i)
find variable features, (ii) reduce dimensions, and (iii) cluster using
both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested_reanalysed</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_nested</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>        <span class="va">data</span>, <span class="op">~</span> <span class="op">{</span></span>
<span>            <span class="va">.x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">.x</span>, subset_row<span class="op">=</span><span class="va">variable_genes</span><span class="op">)</span></span>
<span></span>
<span>            <span class="va">variable_genes</span> <span class="op">&lt;-</span></span>
<span>                <span class="va">.x</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span>prop<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>                <span class="va">.x</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span>use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%$%</span></span>
<span>                <span class="va">membership</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>            <span class="va">.x</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span>ncomponents<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_small_nested_reanalysed</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   cell_class data           </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> lymphoid   <span style="color: #949494;">&lt;SnglCllE[,40]&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> myeloid    <span style="color: #949494;">&lt;SnglCllE[,31]&gt;</span></span></span></code></pre>
<p>We can then unnest and plot the new classification.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested_reanalysed</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Convert to tibble otherwise SingleCellExperiment drops reduced dimensions when unifying data sets.</span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/unnest-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Define unique clusters</span></span>
<span>    <span class="fu"><a href="../reference/unite-methods.html">unite</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cell_class</span>, <span class="va">label</span><span class="op">)</span>, remove<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Plotting</span></span>
<span>    <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>, <span class="va">UMAP2</span>, color<span class="op">=</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">cell_class</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>We can perform a large number of functional analyses on data subsets.
For example, we can identify intra-sample cell-cell interactions using
<em>SingleCellSignalR</em> <span class="citation">(Cabello-Aguilar et
al. 2020)</span>, and then compare whether interactions are stronger or
weaker across conditions. The code below demonstrates how this analysis
could be performed. It won’t work with this small example dataset as we
have just two samples (one for each condition). But some example output
is shown below and you can imagine how you can use tidyverse on the
output to perform t-tests and visualisation.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested_interactions</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_nested_reanalysed</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Unnest based on cell category</span></span>
<span>    <span class="fu"><a href="../reference/unnest-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Create unambiguous clusters</span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span><span class="op">(</span>integrated_clusters<span class="op">=</span><span class="va">first.labels</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span></span>
<span>    <span class="co"># Nest based on sample</span></span>
<span>    <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/nest-methods.html">nest</a></span><span class="op">(</span>data<span class="op">=</span><span class="op">-</span><span class="va">sample</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span><span class="op">(</span>interactions<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="op">{</span></span>
<span></span>
<span>        <span class="co"># Produce variables. Yuck!</span></span>
<span>        <span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">$</span><span class="va">integrated_clusters</span></span>
<span>        <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co"># Ligand/Receptor analysis using SingleCellSignalR</span></span>
<span>        <span class="va">data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/cell_signaling.html" class="external-link">cell_signaling</a></span><span class="op">(</span>genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, cluster<span class="op">=</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/inter_network.html" class="external-link">inter_network</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">data</span>, signal<span class="op">=</span><span class="va">.</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, cluster<span class="op">=</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%$%</span></span>
<span>            <span class="va">`individual-networks`</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="../reference/dplyr-methods.html">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_small_nested_interactions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/unnest-methods.html">unnest</a></span><span class="op">(</span><span class="va">interactions</span><span class="op">)</span></span></code></pre></div>
<p>If the dataset was not so small, and interactions could be
identified, you would see something like below.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tidySingleCellExperiment</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc_small_nested_interactions.html">pbmc_small_nested_interactions</a></span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 100 × 9</span></span></span>
<span><span class="co">##    sample  ligand         recep…¹ ligan…² recep…³ origin desti…⁴ inter…⁵ LRscore</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> sample1 cluster 1.PTMA cluste… PTMA    VIPR1   clust… cluste… paracr…   0.723</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> sample1 cluster 1.B2M  cluste… B2M     KLRD1   clust… cluste… paracr…   0.684</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> sample1 cluster 1.IL16 cluste… IL16    CD4     clust… cluste… paracr…   0.659</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> sample1 cluster 1.HLA… cluste… HLA-B   KLRD1   clust… cluste… paracr…   0.643</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> sample1 cluster 1.CAL… cluste… CALM1   VIPR1   clust… cluste… paracr…   0.616</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> sample1 cluster 1.HLA… cluste… HLA-E   KLRD1   clust… cluste… paracr…   0.585</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> sample1 cluster 1.GNAS cluste… GNAS    VIPR1   clust… cluste… paracr…   0.582</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> sample1 cluster 1.B2M  cluste… B2M     HFE     clust… cluste… paracr…   0.548</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> sample1 cluster 1.PTMA cluste… PTMA    VIPR1   clust… cluste… paracr…   0.704</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> sample1 cluster 1.CAL… cluste… CALM1   VIPR1   clust… cluste… paracr…   0.594</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 90 more rows, and abbreviated variable names ¹​receptor, ²​ligand.name,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ³​receptor.name, ⁴​destination, ⁵​interaction.type</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] tidySingleCellExperiment_1.7.5 ttservice_0.2.2               </span></span>
<span><span class="co">##  [3] tidyHeatmap_1.8.1              purrr_0.3.4                   </span></span>
<span><span class="co">##  [5] SingleCellSignalR_1.8.0        SingleR_1.10.0                </span></span>
<span><span class="co">##  [7] scran_1.24.0                   scater_1.24.0                 </span></span>
<span><span class="co">##  [9] ggplot2_3.3.6                  scuttle_1.6.3                 </span></span>
<span><span class="co">## [11] SingleCellExperiment_1.18.0    SummarizedExperiment_1.26.1   </span></span>
<span><span class="co">## [13] Biobase_2.56.0                 GenomicRanges_1.48.0          </span></span>
<span><span class="co">## [15] GenomeInfoDb_1.32.3            IRanges_2.30.1                </span></span>
<span><span class="co">## [17] S4Vectors_0.34.0               BiocGenerics_0.42.0           </span></span>
<span><span class="co">## [19] MatrixGenerics_1.8.1           matrixStats_0.62.0            </span></span>
<span><span class="co">## [21] knitr_1.40                     BiocStyle_2.24.0              </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] circlize_0.4.15           systemfonts_1.0.4        </span></span>
<span><span class="co">##   [3] plyr_1.8.7                igraph_1.3.4             </span></span>
<span><span class="co">##   [5] lazyeval_0.2.2            splines_4.2.1            </span></span>
<span><span class="co">##   [7] BiocParallel_1.30.3       digest_0.6.29            </span></span>
<span><span class="co">##   [9] foreach_1.5.2             htmltools_0.5.3          </span></span>
<span><span class="co">##  [11] viridis_0.6.2             fansi_1.0.3              </span></span>
<span><span class="co">##  [13] magrittr_2.0.3            memoise_2.0.1            </span></span>
<span><span class="co">##  [15] ScaledMatrix_1.4.0        cluster_2.1.4            </span></span>
<span><span class="co">##  [17] doParallel_1.0.17         limma_3.52.2             </span></span>
<span><span class="co">##  [19] ComplexHeatmap_2.12.1     pkgdown_2.0.6            </span></span>
<span><span class="co">##  [21] colorspace_2.0-3          ggrepel_0.9.1            </span></span>
<span><span class="co">##  [23] textshaping_0.3.6         xfun_0.32                </span></span>
<span><span class="co">##  [25] dplyr_1.0.9               crayon_1.5.1             </span></span>
<span><span class="co">##  [27] RCurl_1.98-1.8            jsonlite_1.8.0           </span></span>
<span><span class="co">##  [29] survival_3.4-0            iterators_1.0.14         </span></span>
<span><span class="co">##  [31] glue_1.6.2                gtable_0.3.0             </span></span>
<span><span class="co">##  [33] zlibbioc_1.42.0           XVector_0.36.0           </span></span>
<span><span class="co">##  [35] SIMLR_1.22.0              GetoptLong_1.0.5         </span></span>
<span><span class="co">##  [37] DelayedArray_0.22.0       BiocSingular_1.12.0      </span></span>
<span><span class="co">##  [39] shape_1.4.6               scales_1.2.1             </span></span>
<span><span class="co">##  [41] pheatmap_1.0.12           GGally_2.1.2             </span></span>
<span><span class="co">##  [43] DBI_1.1.3                 edgeR_3.38.4             </span></span>
<span><span class="co">##  [45] Rcpp_1.0.9                viridisLite_0.4.1        </span></span>
<span><span class="co">##  [47] clue_0.3-61               dqrng_0.3.0              </span></span>
<span><span class="co">##  [49] rsvd_1.0.5                dittoSeq_1.8.1           </span></span>
<span><span class="co">##  [51] httr_1.4.4                htmlwidgets_1.5.4        </span></span>
<span><span class="co">##  [53] metapod_1.4.0             FNN_1.1.3.1              </span></span>
<span><span class="co">##  [55] gplots_3.1.3              RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [57] ellipsis_0.3.2            reshape_0.8.9            </span></span>
<span><span class="co">##  [59] farver_2.1.1              pkgconfig_2.0.3          </span></span>
<span><span class="co">##  [61] uwot_0.1.14               sass_0.4.2               </span></span>
<span><span class="co">##  [63] locfit_1.5-9.6            utf8_1.2.2               </span></span>
<span><span class="co">##  [65] labeling_0.4.2            tidyselect_1.1.2         </span></span>
<span><span class="co">##  [67] rlang_1.0.4               munsell_0.5.0            </span></span>
<span><span class="co">##  [69] tools_4.2.1               cachem_1.0.6             </span></span>
<span><span class="co">##  [71] cli_3.3.0                 generics_0.1.3           </span></span>
<span><span class="co">##  [73] ggridges_0.5.3            evaluate_0.16            </span></span>
<span><span class="co">##  [75] stringr_1.4.1             fastmap_1.1.0            </span></span>
<span><span class="co">##  [77] yaml_2.3.5                ragg_1.2.2               </span></span>
<span><span class="co">##  [79] fs_1.5.2                  caTools_1.18.2           </span></span>
<span><span class="co">##  [81] dendextend_1.16.0         sparseMatrixStats_1.8.0  </span></span>
<span><span class="co">##  [83] pracma_2.3.8              compiler_4.2.1           </span></span>
<span><span class="co">##  [85] plotly_4.10.0             beeswarm_0.4.0           </span></span>
<span><span class="co">##  [87] png_0.1-7                 tibble_3.1.8             </span></span>
<span><span class="co">##  [89] statmod_1.4.37            bslib_0.4.0              </span></span>
<span><span class="co">##  [91] stringi_1.7.8             highr_0.9                </span></span>
<span><span class="co">##  [93] desc_1.4.1                RSpectra_0.16-1          </span></span>
<span><span class="co">##  [95] lattice_0.20-45           bluster_1.6.0            </span></span>
<span><span class="co">##  [97] Matrix_1.4-1              multtest_2.52.0          </span></span>
<span><span class="co">##  [99] vctrs_0.4.1               pillar_1.8.1             </span></span>
<span><span class="co">## [101] lifecycle_1.0.1           BiocManager_1.30.18      </span></span>
<span><span class="co">## [103] jquerylib_0.1.4           GlobalOptions_0.1.2      </span></span>
<span><span class="co">## [105] RcppAnnoy_0.0.19          BiocNeighbors_1.14.0     </span></span>
<span><span class="co">## [107] cowplot_1.1.1             data.table_1.14.2        </span></span>
<span><span class="co">## [109] bitops_1.0-7              irlba_2.3.5              </span></span>
<span><span class="co">## [111] patchwork_1.1.2           R6_2.5.1                 </span></span>
<span><span class="co">## [113] bookdown_0.28             KernSmooth_2.23-20       </span></span>
<span><span class="co">## [115] gridExtra_2.3             vipor_0.4.5              </span></span>
<span><span class="co">## [117] codetools_0.2-18          MASS_7.3-58.1            </span></span>
<span><span class="co">## [119] gtools_3.9.3              assertthat_0.2.1         </span></span>
<span><span class="co">## [121] rprojroot_2.0.3           rjson_0.2.21             </span></span>
<span><span class="co">## [123] withr_2.5.0               GenomeInfoDbData_1.2.8   </span></span>
<span><span class="co">## [125] parallel_4.2.1            grid_4.2.1               </span></span>
<span><span class="co">## [127] beachmat_2.12.0           tidyr_1.2.0              </span></span>
<span><span class="co">## [129] rmarkdown_2.16            DelayedMatrixStats_1.18.0</span></span>
<span><span class="co">## [131] Rtsne_0.16                ggbeeswarm_0.6.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-amezquita2019orchestrating" class="csl-entry">
Amezquita, Robert A, Aaron TL Lun, Etienne Becht, Vince J Carey, Lindsay
N Carpp, Ludwig Geistlinger, Federico Martini, et al. 2019.
<span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span>
<em>Nature Methods</em>, 1–9.
</div>
<div id="ref-aran2019reference" class="csl-entry">
Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong,
Austin Hsu, Suzanna Chak, et al. 2019. <span>“Reference-Based Analysis
of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic
Macrophage.”</span> <em>Nature Immunology</em> 20 (2): 163–72.
</div>
<div id="ref-cabello2020singlecellsignalr" class="csl-entry">
Cabello-Aguilar, Simon, Mélissa Alame, Fabien Kon-Sun-Tack, Caroline
Fau, Matthieu Lacroix, and Jacques Colinge. 2020.
<span>“SingleCellSignalR: Inference of Intercellular Networks from
Single-Cell Featureomics.”</span> <em>Nucleic Acids Research</em> 48
(10): e55–55.
</div>
<div id="ref-lun2016pooling" class="csl-entry">
Lun, Aaron TL, Karsten Bach, and John C Marioni. 2016. <span>“Pooling
Across Cells to Normalize Single-Cell RNA Sequencing Data with Many Zero
Counts.”</span> <em>Genome Biology</em> 17 (1): 75.
</div>
<div id="ref-mangiola2020tidyheatmap" class="csl-entry">
Mangiola, Stefano, and Anthony T Papenfuss. 2020. <span>“tidyHeatmap: An
r Package for Modular Heatmap Production Based on Tidy
Principles.”</span> <em>Journal of Open Source Software</em> 5 (52):
2472.
</div>
<div id="ref-mccarthy2017scater" class="csl-entry">
McCarthy, Davis J, Kieran R Campbell, Aaron TL Lun, and Quin F Wills.
2017. <span>“Scater: Pre-Processing, Quality Control, Normalization and
Visualization of Single-Cell RNA-Seq Data in r.”</span>
<em>Bioinformatics</em> 33 (8): 1179–86.
</div>
<div id="ref-wickham2019welcome" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy
D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019.
<span>“Welcome to the Tidyverse.”</span> <em>Journal of Open Source
Software</em> 4 (43): 1686.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
